{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {
    "id": "k_sHBJLwE6-s"
   },
   "source": [
    "# JoJoGAN: One Shot Face Stylization"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {
    "colab": {
     "base_uri": "https://localhost:8080/",
     "height": 500
    },
    "id": "d1syfTVjLlf7",
    "outputId": "2b1d942e-e9b6-44b3-8337-5d859c27b6ce"
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Requirement already satisfied: pip in d:\\anaconda\\lib\\site-packages (24.2)\n",
      "Requirement already satisfied: setuptools in d:\\anaconda\\lib\\site-packages (74.0.0)\n",
      "Note: you may need to restart the kernel to use updated packages.\n"
     ]
    }
   ],
   "source": [
    "pip install --upgrade pip setuptools"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {
    "colab": {
     "base_uri": "https://localhost:8080/"
    },
    "id": "4Gd6cvXyLoDK",
    "outputId": "abecb268-bdaa-444b-e9b5-7ae2e3c0f931"
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Collecting numpy==1.21.0\n",
      "  Downloading numpy-1.21.0.zip (10.3 MB)\n",
      "     ---------------------------------------- 0.0/10.3 MB ? eta -:--:--\n",
      "     - -------------------------------------- 0.3/10.3 MB ? eta -:--:--\n",
      "     -- ------------------------------------- 0.5/10.3 MB 1.1 MB/s eta 0:00:09\n",
      "     --- ------------------------------------ 0.8/10.3 MB 1.2 MB/s eta 0:00:09\n",
      "     ----- ---------------------------------- 1.3/10.3 MB 1.4 MB/s eta 0:00:07\n",
      "     ------ --------------------------------- 1.6/10.3 MB 1.4 MB/s eta 0:00:07\n",
      "     ------- -------------------------------- 1.8/10.3 MB 1.6 MB/s eta 0:00:06\n",
      "     --------- ------------------------------ 2.4/10.3 MB 1.7 MB/s eta 0:00:05\n",
      "     ----------- ---------------------------- 2.9/10.3 MB 1.8 MB/s eta 0:00:05\n",
      "     ------------- -------------------------- 3.4/10.3 MB 1.8 MB/s eta 0:00:04\n",
      "     --------------- ------------------------ 3.9/10.3 MB 1.8 MB/s eta 0:00:04\n",
      "     ---------------- ----------------------- 4.2/10.3 MB 1.8 MB/s eta 0:00:04\n",
      "     ----------------- ---------------------- 4.5/10.3 MB 1.7 MB/s eta 0:00:04\n",
      "     ------------------ --------------------- 4.7/10.3 MB 1.8 MB/s eta 0:00:04\n",
      "     ------------------ --------------------- 4.7/10.3 MB 1.8 MB/s eta 0:00:04\n",
      "     ------------------ --------------------- 4.7/10.3 MB 1.8 MB/s eta 0:00:04\n",
      "     ------------------ --------------------- 4.7/10.3 MB 1.8 MB/s eta 0:00:04\n",
      "     ------------------ --------------------- 4.7/10.3 MB 1.8 MB/s eta 0:00:04\n",
      "     ------------------ --------------------- 4.7/10.3 MB 1.8 MB/s eta 0:00:04\n",
      "     ------------------ --------------------- 4.7/10.3 MB 1.8 MB/s eta 0:00:04\n",
      "     ------------------ --------------------- 4.7/10.3 MB 1.8 MB/s eta 0:00:04\n",
      "     ------------------ --------------------- 4.7/10.3 MB 1.8 MB/s eta 0:00:04\n",
      "     ------------------ --------------------- 4.7/10.3 MB 1.8 MB/s eta 0:00:04\n",
      "     ------------------ --------------------- 4.7/10.3 MB 1.8 MB/s eta 0:00:04\n",
      "     ------------------ --------------------- 4.7/10.3 MB 1.8 MB/s eta 0:00:04\n",
      "     ------------------ -------------------- 5.0/10.3 MB 920.7 kB/s eta 0:00:06\n",
      "     ------------------ -------------------- 5.0/10.3 MB 920.7 kB/s eta 0:00:06\n",
      "     ------------------- ------------------- 5.2/10.3 MB 923.9 kB/s eta 0:00:06\n",
      "     -------------------- ------------------ 5.5/10.3 MB 924.3 kB/s eta 0:00:06\n",
      "     --------------------- ----------------- 5.8/10.3 MB 929.6 kB/s eta 0:00:05\n",
      "     ---------------------- ---------------- 6.0/10.3 MB 936.7 kB/s eta 0:00:05\n",
      "     -------------------------- ------------- 6.8/10.3 MB 1.0 MB/s eta 0:00:04\n",
      "     ---------------------------- ----------- 7.3/10.3 MB 1.1 MB/s eta 0:00:03\n",
      "     -------------------------------- ------- 8.4/10.3 MB 1.2 MB/s eta 0:00:02\n",
      "     ------------------------------------- -- 9.7/10.3 MB 1.4 MB/s eta 0:00:01\n",
      "     ---------------------------------------- 10.3/10.3 MB 1.4 MB/s eta 0:00:00\n",
      "  Installing build dependencies: started\n",
      "  Installing build dependencies: finished with status 'done'\n",
      "  Getting requirements to build wheel: started\n",
      "  Getting requirements to build wheel: finished with status 'done'\n",
      "Note: you may need to restart the kernel to use updated packages.\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "ERROR: Exception:\n",
      "Traceback (most recent call last):\n",
      "  File \"D:\\anaconda\\Lib\\site-packages\\pip\\_internal\\cli\\base_command.py\", line 105, in _run_wrapper\n",
      "    status = _inner_run()\n",
      "             ^^^^^^^^^^^^\n",
      "  File \"D:\\anaconda\\Lib\\site-packages\\pip\\_internal\\cli\\base_command.py\", line 96, in _inner_run\n",
      "    return self.run(options, args)\n",
      "           ^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"D:\\anaconda\\Lib\\site-packages\\pip\\_internal\\cli\\req_command.py\", line 67, in wrapper\n",
      "    return func(self, options, args)\n",
      "           ^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"D:\\anaconda\\Lib\\site-packages\\pip\\_internal\\commands\\install.py\", line 379, in run\n",
      "    requirement_set = resolver.resolve(\n",
      "                      ^^^^^^^^^^^^^^^^^\n",
      "  File \"D:\\anaconda\\Lib\\site-packages\\pip\\_internal\\resolution\\resolvelib\\resolver.py\", line 95, in resolve\n",
      "    result = self._result = resolver.resolve(\n",
      "                            ^^^^^^^^^^^^^^^^^\n",
      "  File \"D:\\anaconda\\Lib\\site-packages\\pip\\_vendor\\resolvelib\\resolvers.py\", line 546, in resolve\n",
      "    state = resolution.resolve(requirements, max_rounds=max_rounds)\n",
      "            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"D:\\anaconda\\Lib\\site-packages\\pip\\_vendor\\resolvelib\\resolvers.py\", line 397, in resolve\n",
      "    self._add_to_criteria(self.state.criteria, r, parent=None)\n",
      "  File \"D:\\anaconda\\Lib\\site-packages\\pip\\_vendor\\resolvelib\\resolvers.py\", line 173, in _add_to_criteria\n",
      "    if not criterion.candidates:\n",
      "           ^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"D:\\anaconda\\Lib\\site-packages\\pip\\_vendor\\resolvelib\\structs.py\", line 156, in __bool__\n",
      "    return bool(self._sequence)\n",
      "           ^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"D:\\anaconda\\Lib\\site-packages\\pip\\_internal\\resolution\\resolvelib\\found_candidates.py\", line 174, in __bool__\n",
      "    return any(self)\n",
      "           ^^^^^^^^^\n",
      "  File \"D:\\anaconda\\Lib\\site-packages\\pip\\_internal\\resolution\\resolvelib\\found_candidates.py\", line 162, in <genexpr>\n",
      "    return (c for c in iterator if id(c) not in self._incompatible_ids)\n",
      "           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"D:\\anaconda\\Lib\\site-packages\\pip\\_internal\\resolution\\resolvelib\\found_candidates.py\", line 53, in _iter_built\n",
      "    candidate = func()\n",
      "                ^^^^^^\n",
      "  File \"D:\\anaconda\\Lib\\site-packages\\pip\\_internal\\resolution\\resolvelib\\factory.py\", line 186, in _make_candidate_from_link\n",
      "    base: Optional[BaseCandidate] = self._make_base_candidate_from_link(\n",
      "                                    ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"D:\\anaconda\\Lib\\site-packages\\pip\\_internal\\resolution\\resolvelib\\factory.py\", line 232, in _make_base_candidate_from_link\n",
      "    self._link_candidate_cache[link] = LinkCandidate(\n",
      "                                       ^^^^^^^^^^^^^^\n",
      "  File \"D:\\anaconda\\Lib\\site-packages\\pip\\_internal\\resolution\\resolvelib\\candidates.py\", line 303, in __init__\n",
      "    super().__init__(\n",
      "  File \"D:\\anaconda\\Lib\\site-packages\\pip\\_internal\\resolution\\resolvelib\\candidates.py\", line 158, in __init__\n",
      "    self.dist = self._prepare()\n",
      "                ^^^^^^^^^^^^^^^\n",
      "  File \"D:\\anaconda\\Lib\\site-packages\\pip\\_internal\\resolution\\resolvelib\\candidates.py\", line 235, in _prepare\n",
      "    dist = self._prepare_distribution()\n",
      "           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"D:\\anaconda\\Lib\\site-packages\\pip\\_internal\\resolution\\resolvelib\\candidates.py\", line 314, in _prepare_distribution\n",
      "    return preparer.prepare_linked_requirement(self._ireq, parallel_builds=True)\n",
      "           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"D:\\anaconda\\Lib\\site-packages\\pip\\_internal\\operations\\prepare.py\", line 527, in prepare_linked_requirement\n",
      "    return self._prepare_linked_requirement(req, parallel_builds)\n",
      "           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"D:\\anaconda\\Lib\\site-packages\\pip\\_internal\\operations\\prepare.py\", line 642, in _prepare_linked_requirement\n",
      "    dist = _get_prepared_distribution(\n",
      "           ^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"D:\\anaconda\\Lib\\site-packages\\pip\\_internal\\operations\\prepare.py\", line 72, in _get_prepared_distribution\n",
      "    abstract_dist.prepare_distribution_metadata(\n",
      "  File \"D:\\anaconda\\Lib\\site-packages\\pip\\_internal\\distributions\\sdist.py\", line 56, in prepare_distribution_metadata\n",
      "    self._install_build_reqs(finder)\n",
      "  File \"D:\\anaconda\\Lib\\site-packages\\pip\\_internal\\distributions\\sdist.py\", line 126, in _install_build_reqs\n",
      "    build_reqs = self._get_build_requires_wheel()\n",
      "                 ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"D:\\anaconda\\Lib\\site-packages\\pip\\_internal\\distributions\\sdist.py\", line 103, in _get_build_requires_wheel\n",
      "    return backend.get_requires_for_build_wheel()\n",
      "           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"D:\\anaconda\\Lib\\site-packages\\pip\\_internal\\utils\\misc.py\", line 706, in get_requires_for_build_wheel\n",
      "    return super().get_requires_for_build_wheel(config_settings=cs)\n",
      "           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"D:\\anaconda\\Lib\\site-packages\\pip\\_vendor\\pyproject_hooks\\_impl.py\", line 166, in get_requires_for_build_wheel\n",
      "    return self._call_hook('get_requires_for_build_wheel', {\n",
      "           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"D:\\anaconda\\Lib\\site-packages\\pip\\_vendor\\pyproject_hooks\\_impl.py\", line 321, in _call_hook\n",
      "    raise BackendUnavailable(data.get('traceback', ''))\n",
      "pip._vendor.pyproject_hooks._impl.BackendUnavailable: Traceback (most recent call last):\n",
      "  File \"D:\\anaconda\\Lib\\site-packages\\pip\\_vendor\\pyproject_hooks\\_in_process\\_in_process.py\", line 77, in _build_backend\n",
      "    obj = import_module(mod_path)\n",
      "          ^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"D:\\anaconda\\Lib\\importlib\\__init__.py\", line 90, in import_module\n",
      "    return _bootstrap._gcd_import(name[level:], package, level)\n",
      "           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"<frozen importlib._bootstrap>\", line 1387, in _gcd_import\n",
      "  File \"<frozen importlib._bootstrap>\", line 1360, in _find_and_load\n",
      "  File \"<frozen importlib._bootstrap>\", line 1310, in _find_and_load_unlocked\n",
      "  File \"<frozen importlib._bootstrap>\", line 488, in _call_with_frames_removed\n",
      "  File \"<frozen importlib._bootstrap>\", line 1387, in _gcd_import\n",
      "  File \"<frozen importlib._bootstrap>\", line 1360, in _find_and_load\n",
      "  File \"<frozen importlib._bootstrap>\", line 1331, in _find_and_load_unlocked\n",
      "  File \"<frozen importlib._bootstrap>\", line 935, in _load_unlocked\n",
      "  File \"<frozen importlib._bootstrap_external>\", line 995, in exec_module\n",
      "  File \"<frozen importlib._bootstrap>\", line 488, in _call_with_frames_removed\n",
      "  File \"C:\\Users\\Harry\\AppData\\Local\\Temp\\pip-build-env-ihv3gvz4\\overlay\\Lib\\site-packages\\setuptools\\__init__.py\", line 18, in <module>\n",
      "    from setuptools.extern.six import PY3, string_types\n",
      "ModuleNotFoundError: No module named 'setuptools.extern.six'\n",
      "\n"
     ]
    }
   ],
   "source": [
    "pip install numpy==1.21.0\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Collecting contourpyNote: you may need to restart the kernel to use updated packages.\n",
      "\n",
      "  Downloading contourpy-1.3.0-cp312-cp312-win_amd64.whl.metadata (5.4 kB)\n",
      "Collecting numba\n",
      "  Downloading numba-0.60.0-cp312-cp312-win_amd64.whl.metadata (2.8 kB)\n",
      "Collecting pywavelets\n",
      "  Downloading pywavelets-1.7.0-cp312-cp312-win_amd64.whl.metadata (9.0 kB)\n",
      "Collecting streamlit\n",
      "  Downloading streamlit-1.38.0-py2.py3-none-any.whl.metadata (8.5 kB)\n",
      "Collecting numpy>=1.23 (from contourpy)\n",
      "  Using cached numpy-2.1.0-cp312-cp312-win_amd64.whl.metadata (59 kB)\n",
      "Collecting llvmlite<0.44,>=0.43.0dev0 (from numba)\n",
      "  Downloading llvmlite-0.43.0-cp312-cp312-win_amd64.whl.metadata (4.9 kB)\n",
      "Collecting numpy>=1.23 (from contourpy)\n",
      "  Downloading numpy-2.0.2-cp312-cp312-win_amd64.whl.metadata (59 kB)\n",
      "Collecting altair<6,>=4.0 (from streamlit)\n",
      "  Downloading altair-5.4.1-py3-none-any.whl.metadata (9.4 kB)\n",
      "Collecting blinker<2,>=1.0.0 (from streamlit)\n",
      "  Downloading blinker-1.8.2-py3-none-any.whl.metadata (1.6 kB)\n",
      "Collecting cachetools<6,>=4.0 (from streamlit)\n",
      "  Downloading cachetools-5.5.0-py3-none-any.whl.metadata (5.3 kB)\n",
      "Collecting click<9,>=7.0 (from streamlit)\n",
      "  Using cached click-8.1.7-py3-none-any.whl.metadata (3.0 kB)\n",
      "Collecting packaging<25,>=20 (from streamlit)\n",
      "  Using cached packaging-24.1-py3-none-any.whl.metadata (3.2 kB)\n",
      "Collecting pandas<3,>=1.3.0 (from streamlit)\n",
      "  Downloading pandas-2.2.2-cp312-cp312-win_amd64.whl.metadata (19 kB)\n",
      "Collecting pillow<11,>=7.1.0 (from streamlit)\n",
      "  Downloading pillow-10.4.0-cp312-cp312-win_amd64.whl.metadata (9.3 kB)\n",
      "Collecting protobuf<6,>=3.20 (from streamlit)\n",
      "  Downloading protobuf-5.27.4-cp310-abi3-win_amd64.whl.metadata (592 bytes)\n",
      "Collecting pyarrow>=7.0 (from streamlit)\n",
      "  Downloading pyarrow-17.0.0-cp312-cp312-win_amd64.whl.metadata (3.4 kB)\n",
      "Collecting requests<3,>=2.27 (from streamlit)\n",
      "  Using cached requests-2.32.3-py3-none-any.whl.metadata (4.6 kB)\n",
      "Collecting rich<14,>=10.14.0 (from streamlit)\n",
      "  Downloading rich-13.8.0-py3-none-any.whl.metadata (18 kB)\n",
      "Collecting tenacity<9,>=8.1.0 (from streamlit)\n",
      "  Downloading tenacity-8.5.0-py3-none-any.whl.metadata (1.2 kB)\n",
      "Collecting toml<2,>=0.10.1 (from streamlit)\n",
      "  Downloading toml-0.10.2-py2.py3-none-any.whl.metadata (7.1 kB)\n",
      "Collecting typing-extensions<5,>=4.3.0 (from streamlit)\n",
      "  Using cached typing_extensions-4.12.2-py3-none-any.whl.metadata (3.0 kB)\n",
      "Collecting gitpython!=3.1.19,<4,>=3.0.7 (from streamlit)\n",
      "  Using cached GitPython-3.1.43-py3-none-any.whl.metadata (13 kB)\n",
      "Collecting pydeck<1,>=0.8.0b4 (from streamlit)\n",
      "  Downloading pydeck-0.9.1-py2.py3-none-any.whl.metadata (4.1 kB)\n",
      "Collecting tornado<7,>=6.0.3 (from streamlit)\n",
      "  Using cached tornado-6.4.1-cp38-abi3-win_amd64.whl.metadata (2.6 kB)\n",
      "Collecting watchdog<5,>=2.1.5 (from streamlit)\n",
      "  Downloading watchdog-4.0.2-py3-none-win_amd64.whl.metadata (38 kB)\n",
      "Collecting jinja2 (from altair<6,>=4.0->streamlit)\n",
      "  Using cached jinja2-3.1.4-py3-none-any.whl.metadata (2.6 kB)\n",
      "Collecting jsonschema>=3.0 (from altair<6,>=4.0->streamlit)\n",
      "  Downloading jsonschema-4.23.0-py3-none-any.whl.metadata (7.9 kB)\n",
      "Collecting narwhals>=1.5.2 (from altair<6,>=4.0->streamlit)\n",
      "  Downloading narwhals-1.5.5-py3-none-any.whl.metadata (5.8 kB)\n",
      "Collecting colorama (from click<9,>=7.0->streamlit)\n",
      "  Using cached colorama-0.4.6-py2.py3-none-any.whl.metadata (17 kB)\n",
      "Collecting gitdb<5,>=4.0.1 (from gitpython!=3.1.19,<4,>=3.0.7->streamlit)\n",
      "  Using cached gitdb-4.0.11-py3-none-any.whl.metadata (1.2 kB)\n",
      "Collecting python-dateutil>=2.8.2 (from pandas<3,>=1.3.0->streamlit)\n",
      "  Using cached python_dateutil-2.9.0.post0-py2.py3-none-any.whl.metadata (8.4 kB)\n",
      "Collecting pytz>=2020.1 (from pandas<3,>=1.3.0->streamlit)\n",
      "  Using cached pytz-2024.1-py2.py3-none-any.whl.metadata (22 kB)\n",
      "Collecting tzdata>=2022.7 (from pandas<3,>=1.3.0->streamlit)\n",
      "  Using cached tzdata-2024.1-py2.py3-none-any.whl.metadata (1.4 kB)\n",
      "Collecting charset-normalizer<4,>=2 (from requests<3,>=2.27->streamlit)\n",
      "  Downloading charset_normalizer-3.3.2-cp312-cp312-win_amd64.whl.metadata (34 kB)\n",
      "Collecting idna<4,>=2.5 (from requests<3,>=2.27->streamlit)\n",
      "  Downloading idna-3.8-py3-none-any.whl.metadata (9.9 kB)\n",
      "Collecting urllib3<3,>=1.21.1 (from requests<3,>=2.27->streamlit)\n",
      "  Using cached urllib3-2.2.2-py3-none-any.whl.metadata (6.4 kB)\n",
      "Collecting certifi>=2017.4.17 (from requests<3,>=2.27->streamlit)\n",
      "  Using cached certifi-2024.7.4-py3-none-any.whl.metadata (2.2 kB)\n",
      "Collecting markdown-it-py>=2.2.0 (from rich<14,>=10.14.0->streamlit)\n",
      "  Using cached markdown_it_py-3.0.0-py3-none-any.whl.metadata (6.9 kB)\n",
      "Collecting pygments<3.0.0,>=2.13.0 (from rich<14,>=10.14.0->streamlit)\n",
      "  Using cached pygments-2.18.0-py3-none-any.whl.metadata (2.5 kB)\n",
      "Collecting smmap<6,>=3.0.1 (from gitdb<5,>=4.0.1->gitpython!=3.1.19,<4,>=3.0.7->streamlit)\n",
      "  Using cached smmap-5.0.1-py3-none-any.whl.metadata (4.3 kB)\n",
      "Collecting MarkupSafe>=2.0 (from jinja2->altair<6,>=4.0->streamlit)\n",
      "  Downloading MarkupSafe-2.1.5-cp312-cp312-win_amd64.whl.metadata (3.1 kB)\n",
      "Collecting attrs>=22.2.0 (from jsonschema>=3.0->altair<6,>=4.0->streamlit)\n",
      "  Downloading attrs-24.2.0-py3-none-any.whl.metadata (11 kB)\n",
      "Collecting jsonschema-specifications>=2023.03.6 (from jsonschema>=3.0->altair<6,>=4.0->streamlit)\n",
      "  Downloading jsonschema_specifications-2023.12.1-py3-none-any.whl.metadata (3.0 kB)\n",
      "Collecting referencing>=0.28.4 (from jsonschema>=3.0->altair<6,>=4.0->streamlit)\n",
      "  Downloading referencing-0.35.1-py3-none-any.whl.metadata (2.8 kB)\n",
      "Collecting rpds-py>=0.7.1 (from jsonschema>=3.0->altair<6,>=4.0->streamlit)\n",
      "  Downloading rpds_py-0.20.0-cp312-none-win_amd64.whl.metadata (4.2 kB)\n",
      "Collecting mdurl~=0.1 (from markdown-it-py>=2.2.0->rich<14,>=10.14.0->streamlit)\n",
      "  Using cached mdurl-0.1.2-py3-none-any.whl.metadata (1.6 kB)\n",
      "Collecting six>=1.5 (from python-dateutil>=2.8.2->pandas<3,>=1.3.0->streamlit)\n",
      "  Using cached six-1.16.0-py2.py3-none-any.whl.metadata (1.8 kB)\n",
      "Downloading contourpy-1.3.0-cp312-cp312-win_amd64.whl (218 kB)\n",
      "Downloading numba-0.60.0-cp312-cp312-win_amd64.whl (2.7 MB)\n",
      "   ---------------------------------------- 0.0/2.7 MB ? eta -:--:--\n",
      "   --- ------------------------------------ 0.3/2.7 MB ? eta -:--:--\n",
      "   ------- -------------------------------- 0.5/2.7 MB 1.4 MB/s eta 0:00:02\n",
      "   --------------- ------------------------ 1.0/2.7 MB 1.6 MB/s eta 0:00:02\n",
      "   --------------------------- ------------ 1.8/2.7 MB 2.1 MB/s eta 0:00:01\n",
      "   ------------------------------ --------- 2.1/2.7 MB 2.0 MB/s eta 0:00:01\n",
      "   ---------------------------------- ----- 2.4/2.7 MB 1.8 MB/s eta 0:00:01\n",
      "   ---------------------------------------- 2.7/2.7 MB 2.0 MB/s eta 0:00:00\n",
      "Downloading pywavelets-1.7.0-cp312-cp312-win_amd64.whl (4.2 MB)\n",
      "   ---------------------------------------- 0.0/4.2 MB ? eta -:--:--\n",
      "   ------- -------------------------------- 0.8/4.2 MB 8.3 MB/s eta 0:00:01\n",
      "   --------- ------------------------------ 1.0/4.2 MB 2.8 MB/s eta 0:00:02\n",
      "   ------------ --------------------------- 1.3/4.2 MB 2.4 MB/s eta 0:00:02\n",
      "   ----------------- ---------------------- 1.8/4.2 MB 2.0 MB/s eta 0:00:02\n",
      "   ------------------- -------------------- 2.1/4.2 MB 1.9 MB/s eta 0:00:02\n",
      "   ------------------------ --------------- 2.6/4.2 MB 2.3 MB/s eta 0:00:01\n",
      "   ----------------------------- ---------- 3.1/4.2 MB 2.3 MB/s eta 0:00:01\n",
      "   ------------------------------------- -- 3.9/4.2 MB 2.4 MB/s eta 0:00:01\n",
      "   ---------------------------------------- 4.2/4.2 MB 2.3 MB/s eta 0:00:00\n",
      "Downloading streamlit-1.38.0-py2.py3-none-any.whl (8.7 MB)\n",
      "   ---------------------------------------- 0.0/8.7 MB ? eta -:--:--\n",
      "   - -------------------------------------- 0.3/8.7 MB ? eta -:--:--\n",
      "   -- ------------------------------------- 0.5/8.7 MB 3.4 MB/s eta 0:00:03\n",
      "   ---- ----------------------------------- 1.0/8.7 MB 2.6 MB/s eta 0:00:03\n",
      "   -------- ------------------------------- 1.8/8.7 MB 2.9 MB/s eta 0:00:03\n",
      "   ---------- ----------------------------- 2.4/8.7 MB 2.6 MB/s eta 0:00:03\n",
      "   ------------- -------------------------- 2.9/8.7 MB 2.5 MB/s eta 0:00:03\n",
      "   --------------- ------------------------ 3.4/8.7 MB 2.5 MB/s eta 0:00:03\n",
      "   ---------------- ----------------------- 3.7/8.7 MB 2.7 MB/s eta 0:00:02\n",
      "   ------------------- -------------------- 4.2/8.7 MB 2.6 MB/s eta 0:00:02\n",
      "   ---------------------- ----------------- 5.0/8.7 MB 2.6 MB/s eta 0:00:02\n",
      "   ----------------------- ---------------- 5.2/8.7 MB 2.5 MB/s eta 0:00:02\n",
      "   -------------------------- ------------- 5.8/8.7 MB 2.4 MB/s eta 0:00:02\n",
      "   ---------------------------- ----------- 6.3/8.7 MB 2.4 MB/s eta 0:00:02\n",
      "   ----------------------------- ---------- 6.6/8.7 MB 2.3 MB/s eta 0:00:01\n",
      "   -------------------------------- ------- 7.1/8.7 MB 2.5 MB/s eta 0:00:01\n",
      "   ---------------------------------- ----- 7.6/8.7 MB 2.3 MB/s eta 0:00:01\n",
      "   ----------------------------------- ---- 7.9/8.7 MB 2.3 MB/s eta 0:00:01\n",
      "   -------------------------------------- - 8.4/8.7 MB 2.3 MB/s eta 0:00:01\n",
      "   ---------------------------------------- 8.7/8.7 MB 2.3 MB/s eta 0:00:00\n",
      "Downloading altair-5.4.1-py3-none-any.whl (658 kB)\n",
      "   ---------------------------------------- 0.0/658.1 kB ? eta -:--:--\n",
      "   ---------------------------------------- 0.0/658.1 kB ? eta -:--:--\n",
      "   ---------------------------------------- 658.1/658.1 kB 6.4 MB/s eta 0:00:00\n",
      "Downloading blinker-1.8.2-py3-none-any.whl (9.5 kB)\n",
      "Downloading cachetools-5.5.0-py3-none-any.whl (9.5 kB)\n",
      "Using cached click-8.1.7-py3-none-any.whl (97 kB)\n",
      "Using cached GitPython-3.1.43-py3-none-any.whl (207 kB)\n",
      "Downloading llvmlite-0.43.0-cp312-cp312-win_amd64.whl (28.1 MB)\n",
      "   ---------------------------------------- 0.0/28.1 MB ? eta -:--:--\n",
      "   ---------------------------------------- 0.3/28.1 MB ? eta -:--:--\n",
      "   - -------------------------------------- 0.8/28.1 MB 5.6 MB/s eta 0:00:05\n",
      "   - -------------------------------------- 1.3/28.1 MB 3.0 MB/s eta 0:00:09\n",
      "   -- ------------------------------------- 1.6/28.1 MB 2.5 MB/s eta 0:00:11\n",
      "   --- ------------------------------------ 2.4/28.1 MB 2.4 MB/s eta 0:00:11\n",
      "   --- ------------------------------------ 2.6/28.1 MB 2.3 MB/s eta 0:00:12\n",
      "   ---- ----------------------------------- 2.9/28.1 MB 2.1 MB/s eta 0:00:12\n",
      "   ---- ----------------------------------- 3.4/28.1 MB 2.1 MB/s eta 0:00:12\n",
      "   ----- ---------------------------------- 3.9/28.1 MB 2.3 MB/s eta 0:00:11\n",
      "   ------ --------------------------------- 4.5/28.1 MB 2.3 MB/s eta 0:00:11\n",
      "   ------- -------------------------------- 5.2/28.1 MB 2.4 MB/s eta 0:00:10\n",
      "   -------- ------------------------------- 5.8/28.1 MB 2.4 MB/s eta 0:00:10\n",
      "   --------- ------------------------------ 6.6/28.1 MB 2.5 MB/s eta 0:00:09\n",
      "   --------- ------------------------------ 6.6/28.1 MB 2.5 MB/s eta 0:00:09\n",
      "   --------- ------------------------------ 6.8/28.1 MB 2.3 MB/s eta 0:00:10\n",
      "   ---------- ----------------------------- 7.3/28.1 MB 2.2 MB/s eta 0:00:10\n",
      "   ----------- ---------------------------- 8.1/28.1 MB 2.4 MB/s eta 0:00:09\n",
      "   ------------- -------------------------- 9.2/28.1 MB 2.5 MB/s eta 0:00:08\n",
      "   -------------- ------------------------- 10.0/28.1 MB 2.6 MB/s eta 0:00:08\n",
      "   --------------- ------------------------ 10.7/28.1 MB 2.6 MB/s eta 0:00:07\n",
      "   ---------------- ----------------------- 11.8/28.1 MB 2.7 MB/s eta 0:00:07\n",
      "   ----------------- ---------------------- 12.3/28.1 MB 2.7 MB/s eta 0:00:06\n",
      "   ------------------- -------------------- 13.4/28.1 MB 2.8 MB/s eta 0:00:06\n",
      "   ------------------- -------------------- 13.9/28.1 MB 2.9 MB/s eta 0:00:05\n",
      "   --------------------- ------------------ 15.2/28.1 MB 3.0 MB/s eta 0:00:05\n",
      "   ----------------------- ---------------- 16.3/28.1 MB 3.0 MB/s eta 0:00:04\n",
      "   ----------------------- ---------------- 16.5/28.1 MB 3.0 MB/s eta 0:00:04\n",
      "   ------------------------ --------------- 17.6/28.1 MB 3.0 MB/s eta 0:00:04\n",
      "   ------------------------- -------------- 18.1/28.1 MB 3.0 MB/s eta 0:00:04\n",
      "   -------------------------- ------------- 18.6/28.1 MB 3.1 MB/s eta 0:00:04\n",
      "   ---------------------------- ----------- 19.9/28.1 MB 3.1 MB/s eta 0:00:03\n",
      "   ------------------------------ --------- 21.5/28.1 MB 3.3 MB/s eta 0:00:03\n",
      "   -------------------------------- ------- 23.1/28.1 MB 3.4 MB/s eta 0:00:02\n",
      "   ---------------------------------- ----- 24.4/28.1 MB 3.4 MB/s eta 0:00:02\n",
      "   ----------------------------------- ---- 25.2/28.1 MB 3.4 MB/s eta 0:00:01\n",
      "   ------------------------------------ --- 26.0/28.1 MB 3.5 MB/s eta 0:00:01\n",
      "   -------------------------------------- - 27.0/28.1 MB 3.5 MB/s eta 0:00:01\n",
      "   ---------------------------------------- 28.1/28.1 MB 3.6 MB/s eta 0:00:00\n",
      "Downloading numpy-2.0.2-cp312-cp312-win_amd64.whl (15.6 MB)\n",
      "   ---------------------------------------- 0.0/15.6 MB ? eta -:--:--\n",
      "   --- ------------------------------------ 1.3/15.6 MB 13.4 MB/s eta 0:00:02\n",
      "   ------- -------------------------------- 2.9/15.6 MB 8.4 MB/s eta 0:00:02\n",
      "   ----------- ---------------------------- 4.5/15.6 MB 7.3 MB/s eta 0:00:02\n",
      "   -------------- ------------------------- 5.8/15.6 MB 6.8 MB/s eta 0:00:02\n",
      "   ------------------ --------------------- 7.1/15.6 MB 6.6 MB/s eta 0:00:02\n",
      "   -------------------- ------------------- 8.1/15.6 MB 6.2 MB/s eta 0:00:02\n",
      "   ----------------------- ---------------- 9.2/15.6 MB 6.1 MB/s eta 0:00:02\n",
      "   -------------------------- ------------- 10.5/15.6 MB 6.5 MB/s eta 0:00:01\n",
      "   ------------------------------ --------- 11.8/15.6 MB 6.4 MB/s eta 0:00:01\n",
      "   ---------------------------------- ----- 13.6/15.6 MB 6.5 MB/s eta 0:00:01\n",
      "   ----------------------------------- ---- 13.9/15.6 MB 6.2 MB/s eta 0:00:01\n",
      "   ----------------------------------- ---- 13.9/15.6 MB 6.2 MB/s eta 0:00:01\n",
      "   ------------------------------------ --- 14.2/15.6 MB 5.2 MB/s eta 0:00:01\n",
      "   ------------------------------------- -- 14.7/15.6 MB 4.9 MB/s eta 0:00:01\n",
      "   -------------------------------------- - 15.2/15.6 MB 5.0 MB/s eta 0:00:01\n",
      "   ---------------------------------------- 15.6/15.6 MB 4.8 MB/s eta 0:00:00\n",
      "Using cached packaging-24.1-py3-none-any.whl (53 kB)\n",
      "Downloading pandas-2.2.2-cp312-cp312-win_amd64.whl (11.5 MB)\n",
      "   ---------------------------------------- 0.0/11.5 MB ? eta -:--:--\n",
      "    --------------------------------------- 0.3/11.5 MB ? eta -:--:--\n",
      "   ----- ---------------------------------- 1.6/11.5 MB 5.6 MB/s eta 0:00:02\n",
      "   -------- ------------------------------- 2.4/11.5 MB 6.1 MB/s eta 0:00:02\n",
      "   ------------ --------------------------- 3.7/11.5 MB 5.7 MB/s eta 0:00:02\n",
      "   ----------------- ---------------------- 5.0/11.5 MB 5.7 MB/s eta 0:00:02\n",
      "   --------------------- ------------------ 6.3/11.5 MB 5.6 MB/s eta 0:00:01\n",
      "   -------------------------- ------------- 7.6/11.5 MB 5.7 MB/s eta 0:00:01\n",
      "   ----------------------------- ---------- 8.4/11.5 MB 5.4 MB/s eta 0:00:01\n",
      "   -------------------------------- ------- 9.4/11.5 MB 5.3 MB/s eta 0:00:01\n",
      "   ------------------------------------- -- 10.7/11.5 MB 5.8 MB/s eta 0:00:01\n",
      "   ---------------------------------------- 11.5/11.5 MB 5.4 MB/s eta 0:00:00\n",
      "Downloading pillow-10.4.0-cp312-cp312-win_amd64.whl (2.6 MB)\n",
      "   ---------------------------------------- 0.0/2.6 MB ? eta -:--:--\n",
      "   ------------------------ --------------- 1.6/2.6 MB 16.8 MB/s eta 0:00:01\n",
      "   ---------------------------------------- 2.6/2.6 MB 7.7 MB/s eta 0:00:00\n",
      "Downloading protobuf-5.27.4-cp310-abi3-win_amd64.whl (426 kB)\n",
      "Downloading pyarrow-17.0.0-cp312-cp312-win_amd64.whl (25.1 MB)\n",
      "   ---------------------------------------- 0.0/25.1 MB ? eta -:--:--\n",
      "   - -------------------------------------- 1.0/25.1 MB 16.7 MB/s eta 0:00:02\n",
      "   ---- ----------------------------------- 2.6/25.1 MB 7.9 MB/s eta 0:00:03\n",
      "   ------ --------------------------------- 4.2/25.1 MB 7.2 MB/s eta 0:00:03\n",
      "   --------- ------------------------------ 5.8/25.1 MB 6.9 MB/s eta 0:00:03\n",
      "   ----------- ---------------------------- 7.1/25.1 MB 6.6 MB/s eta 0:00:03\n",
      "   ------------ --------------------------- 8.1/25.1 MB 6.4 MB/s eta 0:00:03\n",
      "   -------------- ------------------------- 9.2/25.1 MB 6.3 MB/s eta 0:00:03\n",
      "   ---------------- ----------------------- 10.5/25.1 MB 6.7 MB/s eta 0:00:03\n",
      "   ------------------ --------------------- 11.5/25.1 MB 6.3 MB/s eta 0:00:03\n",
      "   -------------------- ------------------- 13.1/25.1 MB 6.4 MB/s eta 0:00:02\n",
      "   ----------------------- ---------------- 14.9/25.1 MB 6.5 MB/s eta 0:00:02\n",
      "   ------------------------- -------------- 16.0/25.1 MB 6.3 MB/s eta 0:00:02\n",
      "   --------------------------- ------------ 17.0/25.1 MB 6.2 MB/s eta 0:00:02\n",
      "   ---------------------------- ----------- 18.1/25.1 MB 6.1 MB/s eta 0:00:02\n",
      "   ------------------------------ --------- 19.4/25.1 MB 6.4 MB/s eta 0:00:01\n",
      "   --------------------------------- ------ 21.0/25.1 MB 6.4 MB/s eta 0:00:01\n",
      "   ------------------------------------ --- 22.8/25.1 MB 6.4 MB/s eta 0:00:01\n",
      "   -------------------------------------- - 24.4/25.1 MB 6.4 MB/s eta 0:00:01\n",
      "   ---------------------------------------- 25.1/25.1 MB 6.2 MB/s eta 0:00:00\n",
      "Downloading pydeck-0.9.1-py2.py3-none-any.whl (6.9 MB)\n",
      "   ---------------------------------------- 0.0/6.9 MB ? eta -:--:--\n",
      "   ---------- ----------------------------- 1.8/6.9 MB 16.9 MB/s eta 0:00:01\n",
      "   --------------------- ------------------ 3.7/6.9 MB 9.9 MB/s eta 0:00:01\n",
      "   ---------------------------- ----------- 5.0/6.9 MB 8.2 MB/s eta 0:00:01\n",
      "   ------------------------------------- -- 6.6/6.9 MB 7.6 MB/s eta 0:00:01\n",
      "   ---------------------------------------- 6.9/6.9 MB 6.6 MB/s eta 0:00:00\n",
      "Using cached requests-2.32.3-py3-none-any.whl (64 kB)\n",
      "Downloading rich-13.8.0-py3-none-any.whl (241 kB)\n",
      "Downloading tenacity-8.5.0-py3-none-any.whl (28 kB)\n",
      "Downloading toml-0.10.2-py2.py3-none-any.whl (16 kB)\n",
      "Using cached tornado-6.4.1-cp38-abi3-win_amd64.whl (438 kB)\n",
      "Using cached typing_extensions-4.12.2-py3-none-any.whl (37 kB)\n",
      "Downloading watchdog-4.0.2-py3-none-win_amd64.whl (82 kB)\n",
      "Using cached certifi-2024.7.4-py3-none-any.whl (162 kB)\n",
      "Downloading charset_normalizer-3.3.2-cp312-cp312-win_amd64.whl (100 kB)\n",
      "Using cached gitdb-4.0.11-py3-none-any.whl (62 kB)\n",
      "Downloading idna-3.8-py3-none-any.whl (66 kB)\n",
      "Using cached jinja2-3.1.4-py3-none-any.whl (133 kB)\n",
      "Downloading jsonschema-4.23.0-py3-none-any.whl (88 kB)\n",
      "Using cached markdown_it_py-3.0.0-py3-none-any.whl (87 kB)\n",
      "Downloading narwhals-1.5.5-py3-none-any.whl (152 kB)\n",
      "Using cached pygments-2.18.0-py3-none-any.whl (1.2 MB)\n",
      "Using cached python_dateutil-2.9.0.post0-py2.py3-none-any.whl (229 kB)\n",
      "Using cached pytz-2024.1-py2.py3-none-any.whl (505 kB)\n",
      "Using cached tzdata-2024.1-py2.py3-none-any.whl (345 kB)\n",
      "Using cached urllib3-2.2.2-py3-none-any.whl (121 kB)\n",
      "Using cached colorama-0.4.6-py2.py3-none-any.whl (25 kB)\n",
      "Downloading attrs-24.2.0-py3-none-any.whl (63 kB)\n",
      "Downloading jsonschema_specifications-2023.12.1-py3-none-any.whl (18 kB)\n",
      "Downloading MarkupSafe-2.1.5-cp312-cp312-win_amd64.whl (17 kB)\n",
      "Using cached mdurl-0.1.2-py3-none-any.whl (10.0 kB)\n",
      "Downloading referencing-0.35.1-py3-none-any.whl (26 kB)\n",
      "Downloading rpds_py-0.20.0-cp312-none-win_amd64.whl (214 kB)\n",
      "Using cached six-1.16.0-py2.py3-none-any.whl (11 kB)\n",
      "Using cached smmap-5.0.1-py3-none-any.whl (24 kB)\n",
      "Installing collected packages: pytz, watchdog, urllib3, tzdata, typing-extensions, tornado, toml, tenacity, smmap, six, rpds-py, pygments, protobuf, pillow, packaging, numpy, narwhals, mdurl, MarkupSafe, llvmlite, idna, colorama, charset-normalizer, certifi, cachetools, blinker, attrs, requests, referencing, pywavelets, python-dateutil, pyarrow, numba, markdown-it-py, jinja2, gitdb, contourpy, click, rich, pydeck, pandas, jsonschema-specifications, gitpython, jsonschema, altair, streamlit\n",
      "  Attempting uninstall: pytz\n",
      "    Found existing installation: pytz 2024.1\n",
      "    Uninstalling pytz-2024.1:\n",
      "      Successfully uninstalled pytz-2024.1\n",
      "  Attempting uninstall: watchdog\n",
      "    Found existing installation: watchdog 4.0.1\n",
      "    Uninstalling watchdog-4.0.1:\n",
      "      Successfully uninstalled watchdog-4.0.1\n",
      "  Attempting uninstall: urllib3\n",
      "    Found existing installation: urllib3 2.2.2\n",
      "    Uninstalling urllib3-2.2.2:\n",
      "      Successfully uninstalled urllib3-2.2.2\n",
      "  Attempting uninstall: tzdata\n",
      "    Found existing installation: tzdata 2023.3\n",
      "    Uninstalling tzdata-2023.3:\n",
      "      Successfully uninstalled tzdata-2023.3\n",
      "  Attempting uninstall: typing-extensions\n",
      "    Found existing installation: typing_extensions 4.11.0\n",
      "    Uninstalling typing_extensions-4.11.0:\n",
      "      Successfully uninstalled typing_extensions-4.11.0\n",
      "  Attempting uninstall: tornado\n",
      "    Found existing installation: tornado 6.4.1\n",
      "    Uninstalling tornado-6.4.1:\n",
      "      Successfully uninstalled tornado-6.4.1\n",
      "  Attempting uninstall: toml\n",
      "    Found existing installation: toml 0.10.2\n",
      "    Uninstalling toml-0.10.2:\n",
      "      Successfully uninstalled toml-0.10.2\n",
      "  Attempting uninstall: tenacity\n",
      "    Found existing installation: tenacity 8.2.2\n",
      "    Uninstalling tenacity-8.2.2:\n",
      "      Successfully uninstalled tenacity-8.2.2\n",
      "  Attempting uninstall: smmap\n",
      "    Found existing installation: smmap 4.0.0\n",
      "    Uninstalling smmap-4.0.0:\n",
      "      Successfully uninstalled smmap-4.0.0\n",
      "  Attempting uninstall: six\n",
      "    Found existing installation: six 1.16.0\n",
      "    Uninstalling six-1.16.0:\n",
      "      Successfully uninstalled six-1.16.0\n",
      "  Attempting uninstall: rpds-py\n",
      "    Found existing installation: rpds-py 0.10.6\n",
      "    Uninstalling rpds-py-0.10.6:\n",
      "      Successfully uninstalled rpds-py-0.10.6\n",
      "  Attempting uninstall: pygments\n",
      "    Found existing installation: Pygments 2.18.0\n",
      "    Uninstalling Pygments-2.18.0:\n",
      "      Successfully uninstalled Pygments-2.18.0\n",
      "  Attempting uninstall: protobuf\n",
      "    Found existing installation: protobuf 3.20.3\n",
      "    Uninstalling protobuf-3.20.3:\n",
      "      Successfully uninstalled protobuf-3.20.3\n",
      "  Attempting uninstall: pillow\n",
      "    Found existing installation: pillow 10.3.0\n",
      "    Uninstalling pillow-10.3.0:\n",
      "      Successfully uninstalled pillow-10.3.0\n",
      "  Attempting uninstall: packaging\n",
      "    Found existing installation: packaging 24.1\n",
      "    Uninstalling packaging-24.1:\n",
      "      Successfully uninstalled packaging-24.1\n",
      "  Attempting uninstall: numpy\n",
      "    Found existing installation: numpy 2.1.0\n",
      "    Uninstalling numpy-2.1.0:\n",
      "      Successfully uninstalled numpy-2.1.0\n",
      "  Attempting uninstall: mdurl\n",
      "    Found existing installation: mdurl 0.1.0\n",
      "    Uninstalling mdurl-0.1.0:\n",
      "      Successfully uninstalled mdurl-0.1.0\n",
      "  Attempting uninstall: MarkupSafe\n",
      "    Found existing installation: MarkupSafe 2.1.3\n",
      "    Uninstalling MarkupSafe-2.1.3:\n",
      "      Successfully uninstalled MarkupSafe-2.1.3\n",
      "  Attempting uninstall: llvmlite\n",
      "    Found existing installation: llvmlite 0.42.0\n",
      "    Uninstalling llvmlite-0.42.0:\n",
      "      Successfully uninstalled llvmlite-0.42.0\n",
      "  Attempting uninstall: idna\n",
      "    Found existing installation: idna 3.7\n",
      "    Uninstalling idna-3.7:\n",
      "      Successfully uninstalled idna-3.7\n",
      "  Attempting uninstall: colorama\n",
      "    Found existing installation: colorama 0.4.6\n",
      "    Uninstalling colorama-0.4.6:\n",
      "      Successfully uninstalled colorama-0.4.6\n",
      "  Attempting uninstall: charset-normalizer\n",
      "    Found existing installation: charset-normalizer 2.0.4\n",
      "    Uninstalling charset-normalizer-2.0.4:\n",
      "      Successfully uninstalled charset-normalizer-2.0.4\n",
      "  Attempting uninstall: certifi\n",
      "    Found existing installation: certifi 2024.6.2\n",
      "    Uninstalling certifi-2024.6.2:\n",
      "      Successfully uninstalled certifi-2024.6.2\n",
      "  Attempting uninstall: cachetools\n",
      "    Found existing installation: cachetools 5.3.3\n",
      "    Uninstalling cachetools-5.3.3:\n",
      "      Successfully uninstalled cachetools-5.3.3\n",
      "  Attempting uninstall: blinker\n",
      "    Found existing installation: blinker 1.6.2\n",
      "    Uninstalling blinker-1.6.2:\n",
      "      Successfully uninstalled blinker-1.6.2\n",
      "  Attempting uninstall: attrs\n",
      "    Found existing installation: attrs 23.1.0\n",
      "    Uninstalling attrs-23.1.0:\n",
      "      Successfully uninstalled attrs-23.1.0\n",
      "  Attempting uninstall: requests\n",
      "    Found existing installation: requests 2.32.2\n",
      "    Uninstalling requests-2.32.2:\n",
      "      Successfully uninstalled requests-2.32.2\n",
      "  Attempting uninstall: referencing\n",
      "    Found existing installation: referencing 0.30.2\n",
      "    Uninstalling referencing-0.30.2:\n",
      "      Successfully uninstalled referencing-0.30.2\n",
      "  Attempting uninstall: pywavelets\n",
      "    Found existing installation: pywavelets 1.5.0\n",
      "    Uninstalling pywavelets-1.5.0:\n",
      "      Successfully uninstalled pywavelets-1.5.0\n",
      "  Attempting uninstall: python-dateutil\n",
      "    Found existing installation: python-dateutil 2.9.0.post0\n",
      "    Uninstalling python-dateutil-2.9.0.post0:\n",
      "      Successfully uninstalled python-dateutil-2.9.0.post0\n",
      "  Attempting uninstall: pyarrow\n",
      "    Found existing installation: pyarrow 14.0.2\n",
      "    Uninstalling pyarrow-14.0.2:\n",
      "      Successfully uninstalled pyarrow-14.0.2\n",
      "  Attempting uninstall: numba\n",
      "    Found existing installation: numba 0.59.1\n",
      "    Uninstalling numba-0.59.1:\n",
      "      Successfully uninstalled numba-0.59.1\n",
      "  Attempting uninstall: markdown-it-py\n",
      "    Found existing installation: markdown-it-py 2.2.0\n",
      "    Uninstalling markdown-it-py-2.2.0:\n",
      "      Successfully uninstalled markdown-it-py-2.2.0\n",
      "  Attempting uninstall: jinja2\n",
      "    Found existing installation: Jinja2 3.1.4\n",
      "    Uninstalling Jinja2-3.1.4:\n",
      "      Successfully uninstalled Jinja2-3.1.4\n",
      "  Attempting uninstall: gitdb\n",
      "    Found existing installation: gitdb 4.0.7\n",
      "    Uninstalling gitdb-4.0.7:\n",
      "      Successfully uninstalled gitdb-4.0.7\n",
      "  Attempting uninstall: contourpy\n",
      "    Found existing installation: contourpy 1.2.0\n",
      "    Uninstalling contourpy-1.2.0:\n",
      "      Successfully uninstalled contourpy-1.2.0\n",
      "  Attempting uninstall: click\n",
      "    Found existing installation: click 8.1.7\n",
      "    Uninstalling click-8.1.7:\n",
      "      Successfully uninstalled click-8.1.7\n",
      "  Attempting uninstall: rich\n",
      "    Found existing installation: rich 13.3.5\n",
      "    Uninstalling rich-13.3.5:\n",
      "      Successfully uninstalled rich-13.3.5\n",
      "  Attempting uninstall: pydeck\n",
      "    Found existing installation: pydeck 0.8.0\n",
      "    Uninstalling pydeck-0.8.0:\n",
      "      Successfully uninstalled pydeck-0.8.0\n",
      "  Attempting uninstall: pandas\n",
      "    Found existing installation: pandas 2.2.2\n",
      "    Uninstalling pandas-2.2.2:\n",
      "      Successfully uninstalled pandas-2.2.2\n",
      "  Attempting uninstall: jsonschema-specifications\n",
      "    Found existing installation: jsonschema-specifications 2023.7.1\n",
      "    Uninstalling jsonschema-specifications-2023.7.1:\n",
      "      Successfully uninstalled jsonschema-specifications-2023.7.1\n",
      "  Attempting uninstall: gitpython\n",
      "    Found existing installation: GitPython 3.1.37\n",
      "    Uninstalling GitPython-3.1.37:\n",
      "      Successfully uninstalled GitPython-3.1.37\n",
      "  Attempting uninstall: jsonschema\n",
      "    Found existing installation: jsonschema 4.19.2\n",
      "    Uninstalling jsonschema-4.19.2:\n",
      "      Successfully uninstalled jsonschema-4.19.2\n",
      "  Attempting uninstall: altair\n",
      "    Found existing installation: altair 5.0.1\n",
      "    Uninstalling altair-5.0.1:\n",
      "      Successfully uninstalled altair-5.0.1\n",
      "  Attempting uninstall: streamlit\n",
      "    Found existing installation: streamlit 1.32.0\n",
      "    Uninstalling streamlit-1.32.0:\n",
      "      Successfully uninstalled streamlit-1.32.0\n",
      "Successfully installed MarkupSafe-2.1.5 altair-5.4.1 attrs-24.2.0 blinker-1.8.2 cachetools-5.5.0 certifi-2024.7.4 charset-normalizer-3.3.2 click-8.1.7 colorama-0.4.6 contourpy-1.3.0 gitdb-4.0.11 gitpython-3.1.43 idna-3.8 jinja2-3.1.4 jsonschema-4.23.0 jsonschema-specifications-2023.12.1 llvmlite-0.43.0 markdown-it-py-3.0.0 mdurl-0.1.2 narwhals-1.5.5 numba-0.60.0 numpy-2.0.2 packaging-23.2 pandas-2.2.2 pillow-10.4.0 protobuf-5.27.4 pyarrow-17.0.0 pydeck-0.9.1 pygments-2.15.1 python-dateutil-2.9.0.post0 pytz-2024.1 pywavelets-1.7.0 referencing-0.35.1 requests-2.32.3 rich-13.8.0 rpds-py-0.20.0 six-1.16.0 smmap-5.0.1 streamlit-1.38.0 tenacity-8.5.0 toml-0.10.2 tornado-6.4.1 typing-extensions-4.12.2 tzdata-2024.1 urllib3-2.2.2 watchdog-4.0.2\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "  WARNING: Failed to remove contents in a temporary directory 'C:\\Users\\Harry\\AppData\\Roaming\\Python\\Python312\\site-packages\\~ornado'.\n",
      "  You can safely remove it manually.\n",
      "ERROR: pip's dependency resolver does not currently take into account all the packages that are installed. This behaviour is the source of the following dependency conflicts.\n",
      "mdit-py-plugins 0.3.0 requires markdown-it-py<3.0.0,>=1.0.0, but you have markdown-it-py 3.0.0 which is incompatible.\n"
     ]
    }
   ],
   "source": [
    "pip install --force-reinstall contourpy numba pywavelets streamlit\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "[autoreload of google.protobuf.internal.python_message failed: Traceback (most recent call last):\n",
      "  File \"C:\\Users\\Harry\\AppData\\Roaming\\Python\\Python312\\site-packages\\IPython\\extensions\\autoreload.py\", line 276, in check\n",
      "    superreload(m, reload, self.old_objects)\n",
      "  File \"C:\\Users\\Harry\\AppData\\Roaming\\Python\\Python312\\site-packages\\IPython\\extensions\\autoreload.py\", line 500, in superreload\n",
      "    update_generic(old_obj, new_obj)\n",
      "  File \"C:\\Users\\Harry\\AppData\\Roaming\\Python\\Python312\\site-packages\\IPython\\extensions\\autoreload.py\", line 397, in update_generic\n",
      "    update(a, b)\n",
      "  File \"C:\\Users\\Harry\\AppData\\Roaming\\Python\\Python312\\site-packages\\IPython\\extensions\\autoreload.py\", line 365, in update_class\n",
      "    update_instances(old, new)\n",
      "  File \"C:\\Users\\Harry\\AppData\\Roaming\\Python\\Python312\\site-packages\\IPython\\extensions\\autoreload.py\", line 323, in update_instances\n",
      "    object.__setattr__(ref, \"__class__\", new)\n",
      "TypeError: can't apply this __setattr__ to GeneratedProtocolMessageType object\n",
      "]\n",
      "[autoreload of google.protobuf.descriptor_pb2 failed: Traceback (most recent call last):\n",
      "  File \"C:\\Users\\Harry\\AppData\\Roaming\\Python\\Python312\\site-packages\\IPython\\extensions\\autoreload.py\", line 276, in check\n",
      "    superreload(m, reload, self.old_objects)\n",
      "  File \"C:\\Users\\Harry\\AppData\\Roaming\\Python\\Python312\\site-packages\\IPython\\extensions\\autoreload.py\", line 475, in superreload\n",
      "    module = reload(module)\n",
      "             ^^^^^^^^^^^^^^\n",
      "  File \"D:\\anaconda\\Lib\\importlib\\__init__.py\", line 131, in reload\n",
      "    _bootstrap._exec(spec, module)\n",
      "  File \"<frozen importlib._bootstrap>\", line 866, in _exec\n",
      "  File \"<frozen importlib._bootstrap_external>\", line 995, in exec_module\n",
      "  File \"<frozen importlib._bootstrap>\", line 488, in _call_with_frames_removed\n",
      "  File \"D:\\anaconda\\Lib\\site-packages\\google\\protobuf\\descriptor_pb2.py\", line 3037, in <module>\n",
      "    _builder.BuildTopDescriptorsAndMessages(DESCRIPTOR, 'google.protobuf.descriptor_pb2', _globals)\n",
      "  File \"D:\\anaconda\\Lib\\site-packages\\google\\protobuf\\internal\\builder.py\", line 86, in BuildTopDescriptorsAndMessages\n",
      "    module[name] = BuildMessage(msg_des)\n",
      "                   ^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"D:\\anaconda\\Lib\\site-packages\\google\\protobuf\\internal\\builder.py\", line 63, in BuildMessage\n",
      "    message_class = _reflection.GeneratedProtocolMessageType(\n",
      "                    ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "TypeError: A Message class can only inherit from Message, not (<class 'google.protobuf.message.Message'>,)\n",
      "]\n",
      "[autoreload of charset_normalizer.cd failed: Traceback (most recent call last):\n",
      "  File \"C:\\Users\\Harry\\AppData\\Roaming\\Python\\Python312\\site-packages\\IPython\\extensions\\autoreload.py\", line 276, in check\n",
      "    superreload(m, reload, self.old_objects)\n",
      "  File \"C:\\Users\\Harry\\AppData\\Roaming\\Python\\Python312\\site-packages\\IPython\\extensions\\autoreload.py\", line 475, in superreload\n",
      "    module = reload(module)\n",
      "             ^^^^^^^^^^^^^^\n",
      "  File \"D:\\anaconda\\Lib\\importlib\\__init__.py\", line 131, in reload\n",
      "    _bootstrap._exec(spec, module)\n",
      "  File \"<frozen importlib._bootstrap>\", line 866, in _exec\n",
      "  File \"<frozen importlib._bootstrap_external>\", line 995, in exec_module\n",
      "  File \"<frozen importlib._bootstrap>\", line 488, in _call_with_frames_removed\n",
      "  File \"D:\\anaconda\\Lib\\site-packages\\charset_normalizer\\cd.py\", line 14, in <module>\n",
      "    from .md import is_suspiciously_successive_range\n",
      "ImportError: cannot import name 'is_suspiciously_successive_range' from 'charset_normalizer.md' (D:\\anaconda\\Lib\\site-packages\\charset_normalizer\\md.cp312-win_amd64.pyd)\n",
      "]\n",
      "[autoreload of charset_normalizer.api failed: Traceback (most recent call last):\n",
      "  File \"C:\\Users\\Harry\\AppData\\Roaming\\Python\\Python312\\site-packages\\IPython\\extensions\\autoreload.py\", line 276, in check\n",
      "    superreload(m, reload, self.old_objects)\n",
      "  File \"C:\\Users\\Harry\\AppData\\Roaming\\Python\\Python312\\site-packages\\IPython\\extensions\\autoreload.py\", line 475, in superreload\n",
      "    module = reload(module)\n",
      "             ^^^^^^^^^^^^^^\n",
      "  File \"D:\\anaconda\\Lib\\importlib\\__init__.py\", line 131, in reload\n",
      "    _bootstrap._exec(spec, module)\n",
      "  File \"<frozen importlib._bootstrap>\", line 866, in _exec\n",
      "  File \"<frozen importlib._bootstrap_external>\", line 995, in exec_module\n",
      "  File \"<frozen importlib._bootstrap>\", line 488, in _call_with_frames_removed\n",
      "  File \"D:\\anaconda\\Lib\\site-packages\\charset_normalizer\\api.py\", line 12, in <module>\n",
      "    from .md import mess_ratio\n",
      "ImportError: cannot import name 'mess_ratio' from 'charset_normalizer.md' (D:\\anaconda\\Lib\\site-packages\\charset_normalizer\\md.cp312-win_amd64.pyd)\n",
      "]\n",
      "[autoreload of charset_normalizer failed: Traceback (most recent call last):\n",
      "  File \"C:\\Users\\Harry\\AppData\\Roaming\\Python\\Python312\\site-packages\\IPython\\extensions\\autoreload.py\", line 276, in check\n",
      "    superreload(m, reload, self.old_objects)\n",
      "  File \"C:\\Users\\Harry\\AppData\\Roaming\\Python\\Python312\\site-packages\\IPython\\extensions\\autoreload.py\", line 475, in superreload\n",
      "    module = reload(module)\n",
      "             ^^^^^^^^^^^^^^\n",
      "  File \"D:\\anaconda\\Lib\\importlib\\__init__.py\", line 131, in reload\n",
      "    _bootstrap._exec(spec, module)\n",
      "  File \"<frozen importlib._bootstrap>\", line 866, in _exec\n",
      "  File \"<frozen importlib._bootstrap_external>\", line 995, in exec_module\n",
      "  File \"<frozen importlib._bootstrap>\", line 488, in _call_with_frames_removed\n",
      "  File \"D:\\anaconda\\Lib\\site-packages\\charset_normalizer\\__init__.py\", line 24, in <module>\n",
      "    from .api import from_bytes, from_fp, from_path, is_binary\n",
      "ImportError: cannot import name 'is_binary' from 'charset_normalizer.api' (D:\\anaconda\\Lib\\site-packages\\charset_normalizer\\api.py)\n",
      "]\n"
     ]
    }
   ],
   "source": [
    "pip check\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Collecting scikit-learn==0.22\n",
      "  Using cached scikit-learn-0.22.tar.gz (6.9 MB)\n",
      "  Preparing metadata (setup.py): started\n",
      "  Preparing metadata (setup.py): finished with status 'done'\n",
      "Requirement already satisfied: numpy>=1.11.0 in d:\\anaconda\\lib\\site-packages (from scikit-learn==0.22) (2.1.0)\n",
      "Requirement already satisfied: scipy>=0.17.0 in d:\\anaconda\\lib\\site-packages (from scikit-learn==0.22) (1.13.1)\n",
      "Requirement already satisfied: joblib>=0.11 in d:\\anaconda\\lib\\site-packages (from scikit-learn==0.22) (1.4.2)\n",
      "Building wheels for collected packages: scikit-learn\n",
      "  Building wheel for scikit-learn (setup.py): started\n",
      "  Building wheel for scikit-learn (setup.py): finished with status 'error'\n",
      "  Running setup.py clean for scikit-learn\n",
      "Failed to build scikit-learn\n",
      "Note: you may need to restart the kernel to use updated packages.\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "  error: subprocess-exited-with-error\n",
      "  \n",
      "  python setup.py bdist_wheel did not run successfully.\n",
      "  exit code: 1\n",
      "  \n",
      "  [11 lines of output]\n",
      "  C:\\Users\\Harry\\AppData\\Local\\Temp\\pip-install-drnmec38\\scikit-learn_0516d3f7c78a48e385f1495be3ffe719\\setup.py:12: DeprecationWarning: pkg_resources is deprecated as an API. See https://setuptools.pypa.io/en/latest/pkg_resources.html\n",
      "    from pkg_resources import parse_version\n",
      "  Partial import of sklearn during the build process.\n",
      "  Traceback (most recent call last):\n",
      "    File \"<string>\", line 2, in <module>\n",
      "    File \"<pip-setuptools-caller>\", line 34, in <module>\n",
      "    File \"C:\\Users\\Harry\\AppData\\Local\\Temp\\pip-install-drnmec38\\scikit-learn_0516d3f7c78a48e385f1495be3ffe719\\setup.py\", line 303, in <module>\n",
      "      setup_package()\n",
      "    File \"C:\\Users\\Harry\\AppData\\Local\\Temp\\pip-install-drnmec38\\scikit-learn_0516d3f7c78a48e385f1495be3ffe719\\setup.py\", line 295, in setup_package\n",
      "      from numpy.distutils.core import setup\n",
      "  ModuleNotFoundError: No module named 'numpy.distutils'\n",
      "  [end of output]\n",
      "  \n",
      "  note: This error originates from a subprocess, and is likely not a problem with pip.\n",
      "  ERROR: Failed building wheel for scikit-learn\n",
      "ERROR: ERROR: Failed to build installable wheels for some pyproject.toml based projects (scikit-learn)\n"
     ]
    }
   ],
   "source": [
    "pip install scikit-learn==0.22 --no-build-isolation --no-use-pep517"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Requirement already satisfied: opencv-python in d:\\anaconda\\lib\\site-packages (4.10.0.84)\n",
      "Requirement already satisfied: numpy>=1.21.2 in d:\\anaconda\\lib\\site-packages (from opencv-python) (2.1.0)\n",
      "Note: you may need to restart the kernel to use updated packages.\n"
     ]
    }
   ],
   "source": [
    "pip install opencv-python\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Requirement already satisfied: cmake in d:\\anaconda\\lib\\site-packages (3.30.2)\n",
      "Note: you may need to restart the kernel to use updated packages.\n"
     ]
    }
   ],
   "source": [
    "pip install cmake"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Requirement already satisfied: dlib in d:\\anaconda\\lib\\site-packages (19.24.6)\n",
      "Note: you may need to restart the kernel to use updated packages.\n"
     ]
    }
   ],
   "source": [
    "pip install dlib"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Looking in indexes: https://download.pytorch.org/whl/cu118\n",
      "Requirement already satisfied: torch in d:\\anaconda\\lib\\site-packages (2.4.0+cu118)\n",
      "Requirement already satisfied: torchvision in d:\\anaconda\\lib\\site-packages (0.19.0+cu118)\n",
      "Requirement already satisfied: torchaudio in d:\\anaconda\\lib\\site-packages (2.4.0+cu118)\n",
      "Requirement already satisfied: filelock in d:\\anaconda\\lib\\site-packages (from torch) (3.13.1)\n",
      "Requirement already satisfied: typing-extensions>=4.8.0 in d:\\anaconda\\lib\\site-packages (from torch) (4.11.0)\n",
      "Requirement already satisfied: sympy in d:\\anaconda\\lib\\site-packages (from torch) (1.12)\n",
      "Requirement already satisfied: networkx in d:\\anaconda\\lib\\site-packages (from torch) (3.2.1)\n",
      "Requirement already satisfied: jinja2 in d:\\anaconda\\lib\\site-packages (from torch) (3.1.4)\n",
      "Requirement already satisfied: fsspec in d:\\anaconda\\lib\\site-packages (from torch) (2024.3.1)\n",
      "Requirement already satisfied: setuptools in d:\\anaconda\\lib\\site-packages (from torch) (74.0.0)\n",
      "Requirement already satisfied: numpy in d:\\anaconda\\lib\\site-packages (from torchvision) (2.1.0)\n",
      "Requirement already satisfied: pillow!=8.3.*,>=5.3.0 in d:\\anaconda\\lib\\site-packages (from torchvision) (10.3.0)\n",
      "Requirement already satisfied: MarkupSafe>=2.0 in d:\\anaconda\\lib\\site-packages (from jinja2->torch) (2.1.3)\n",
      "Requirement already satisfied: mpmath>=0.19 in d:\\anaconda\\lib\\site-packages (from sympy->torch) (1.3.0)\n",
      "Note: you may need to restart the kernel to use updated packages.\n"
     ]
    }
   ],
   "source": [
    "pip install torch torchvision torchaudio --index-url https://download.pytorch.org/whl/cu118\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Requirement already satisfied: tqdm in d:\\anaconda\\lib\\site-packages (4.66.4)\n",
      "Requirement already satisfied: wandb in d:\\anaconda\\lib\\site-packages (0.17.7)\n",
      "Requirement already satisfied: ninja in d:\\anaconda\\lib\\site-packages (1.11.1.1)\n",
      "Requirement already satisfied: colorama in c:\\users\\harry\\appdata\\roaming\\python\\python312\\site-packages (from tqdm) (0.4.6)\n",
      "Requirement already satisfied: click!=8.0.0,>=7.1 in d:\\anaconda\\lib\\site-packages (from wandb) (8.1.7)\n",
      "Requirement already satisfied: docker-pycreds>=0.4.0 in d:\\anaconda\\lib\\site-packages (from wandb) (0.4.0)\n",
      "Requirement already satisfied: gitpython!=3.1.29,>=1.0.0 in d:\\anaconda\\lib\\site-packages (from wandb) (3.1.37)\n",
      "Requirement already satisfied: platformdirs in c:\\users\\harry\\appdata\\roaming\\python\\python312\\site-packages (from wandb) (4.2.2)\n",
      "Requirement already satisfied: protobuf!=4.21.0,<6,>=3.19.0 in d:\\anaconda\\lib\\site-packages (from wandb) (3.20.3)\n",
      "Requirement already satisfied: psutil>=5.0.0 in c:\\users\\harry\\appdata\\roaming\\python\\python312\\site-packages (from wandb) (6.0.0)\n",
      "Requirement already satisfied: pyyaml in d:\\anaconda\\lib\\site-packages (from wandb) (6.0.1)\n",
      "Requirement already satisfied: requests<3,>=2.0.0 in d:\\anaconda\\lib\\site-packages (from wandb) (2.32.2)\n",
      "Requirement already satisfied: sentry-sdk>=1.0.0 in d:\\anaconda\\lib\\site-packages (from wandb) (2.13.0)\n",
      "Requirement already satisfied: setproctitle in d:\\anaconda\\lib\\site-packages (from wandb) (1.3.3)\n",
      "Requirement already satisfied: setuptools in d:\\anaconda\\lib\\site-packages (from wandb) (74.0.0)\n",
      "Requirement already satisfied: six>=1.4.0 in c:\\users\\harry\\appdata\\roaming\\python\\python312\\site-packages (from docker-pycreds>=0.4.0->wandb) (1.16.0)\n",
      "Requirement already satisfied: gitdb<5,>=4.0.1 in d:\\anaconda\\lib\\site-packages (from gitpython!=3.1.29,>=1.0.0->wandb) (4.0.7)\n",
      "Requirement already satisfied: charset-normalizer<4,>=2 in d:\\anaconda\\lib\\site-packages (from requests<3,>=2.0.0->wandb) (2.0.4)\n",
      "Requirement already satisfied: idna<4,>=2.5 in d:\\anaconda\\lib\\site-packages (from requests<3,>=2.0.0->wandb) (3.7)\n",
      "Requirement already satisfied: urllib3<3,>=1.21.1 in d:\\anaconda\\lib\\site-packages (from requests<3,>=2.0.0->wandb) (2.2.2)\n",
      "Requirement already satisfied: certifi>=2017.4.17 in d:\\anaconda\\lib\\site-packages (from requests<3,>=2.0.0->wandb) (2024.6.2)\n",
      "Requirement already satisfied: smmap<5,>=3.0.1 in d:\\anaconda\\lib\\site-packages (from gitdb<5,>=4.0.1->gitpython!=3.1.29,>=1.0.0->wandb) (4.0.0)\n",
      "Note: you may need to restart the kernel to use updated packages.\n"
     ]
    }
   ],
   "source": [
    "pip install tqdm wandb ninja\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Requirement already satisfied: ninja in d:\\anaconda\\lib\\site-packages (1.11.1.1)\n",
      "Note: you may need to restart the kernel to use updated packages.\n"
     ]
    }
   ],
   "source": [
    "pip install ninja\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {
    "colab": {
     "base_uri": "https://localhost:8080/"
    },
    "id": "81RU0HATLqjc",
    "outputId": "31ea34eb-a979-4f87-eb00-8d65cdaa69ea"
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Requirement already satisfied: wandb in d:\\anaconda\\lib\\site-packages (0.17.7)\n",
      "Requirement already satisfied: click!=8.0.0,>=7.1 in d:\\anaconda\\lib\\site-packages (from wandb) (8.1.7)\n",
      "Requirement already satisfied: docker-pycreds>=0.4.0 in d:\\anaconda\\lib\\site-packages (from wandb) (0.4.0)\n",
      "Requirement already satisfied: gitpython!=3.1.29,>=1.0.0 in d:\\anaconda\\lib\\site-packages (from wandb) (3.1.37)\n",
      "Requirement already satisfied: platformdirs in c:\\users\\harry\\appdata\\roaming\\python\\python312\\site-packages (from wandb) (4.2.2)\n",
      "Requirement already satisfied: protobuf!=4.21.0,<6,>=3.19.0 in d:\\anaconda\\lib\\site-packages (from wandb) (3.20.3)\n",
      "Requirement already satisfied: psutil>=5.0.0 in c:\\users\\harry\\appdata\\roaming\\python\\python312\\site-packages (from wandb) (6.0.0)\n",
      "Requirement already satisfied: pyyaml in d:\\anaconda\\lib\\site-packages (from wandb) (6.0.1)\n",
      "Requirement already satisfied: requests<3,>=2.0.0 in d:\\anaconda\\lib\\site-packages (from wandb) (2.32.2)\n",
      "Requirement already satisfied: sentry-sdk>=1.0.0 in d:\\anaconda\\lib\\site-packages (from wandb) (2.13.0)\n",
      "Requirement already satisfied: setproctitle in d:\\anaconda\\lib\\site-packages (from wandb) (1.3.3)\n",
      "Requirement already satisfied: setuptools in d:\\anaconda\\lib\\site-packages (from wandb) (74.0.0)\n",
      "Requirement already satisfied: colorama in c:\\users\\harry\\appdata\\roaming\\python\\python312\\site-packages (from click!=8.0.0,>=7.1->wandb) (0.4.6)\n",
      "Requirement already satisfied: six>=1.4.0 in c:\\users\\harry\\appdata\\roaming\\python\\python312\\site-packages (from docker-pycreds>=0.4.0->wandb) (1.16.0)\n",
      "Requirement already satisfied: gitdb<5,>=4.0.1 in d:\\anaconda\\lib\\site-packages (from gitpython!=3.1.29,>=1.0.0->wandb) (4.0.7)\n",
      "Requirement already satisfied: charset-normalizer<4,>=2 in d:\\anaconda\\lib\\site-packages (from requests<3,>=2.0.0->wandb) (2.0.4)\n",
      "Requirement already satisfied: idna<4,>=2.5 in d:\\anaconda\\lib\\site-packages (from requests<3,>=2.0.0->wandb) (3.7)\n",
      "Requirement already satisfied: urllib3<3,>=1.21.1 in d:\\anaconda\\lib\\site-packages (from requests<3,>=2.0.0->wandb) (2.2.2)\n",
      "Requirement already satisfied: certifi>=2017.4.17 in d:\\anaconda\\lib\\site-packages (from requests<3,>=2.0.0->wandb) (2024.6.2)\n",
      "Requirement already satisfied: smmap<5,>=3.0.1 in d:\\anaconda\\lib\\site-packages (from gitdb<5,>=4.0.1->gitpython!=3.1.29,>=1.0.0->wandb) (4.0.0)\n",
      "Note: you may need to restart the kernel to use updated packages.\n"
     ]
    }
   ],
   "source": [
    "pip install wandb"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Requirement already satisfied: numpy in d:\\anaconda\\lib\\site-packages (2.1.0)\n",
      "Note: you may need to restart the kernel to use updated packages.\n"
     ]
    }
   ],
   "source": [
    "pip install --upgrade numpy\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Note: you may need to restart the kernel to use updated packages.\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "The syntax of the command is incorrect.\n"
     ]
    }
   ],
   "source": [
    "pip install <path_to_downloaded_wheel>\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Requirement already satisfied: matplotlib in d:\\anaconda\\lib\\site-packages (3.8.4)\n",
      "Requirement already satisfied: contourpy>=1.0.1 in d:\\anaconda\\lib\\site-packages (from matplotlib) (1.2.0)\n",
      "Requirement already satisfied: cycler>=0.10 in d:\\anaconda\\lib\\site-packages (from matplotlib) (0.11.0)\n",
      "Requirement already satisfied: fonttools>=4.22.0 in d:\\anaconda\\lib\\site-packages (from matplotlib) (4.51.0)\n",
      "Requirement already satisfied: kiwisolver>=1.3.1 in d:\\anaconda\\lib\\site-packages (from matplotlib) (1.4.4)\n",
      "Requirement already satisfied: numpy>=1.21 in d:\\anaconda\\lib\\site-packages (from matplotlib) (2.1.0)\n",
      "Requirement already satisfied: packaging>=20.0 in c:\\users\\harry\\appdata\\roaming\\python\\python312\\site-packages (from matplotlib) (24.1)\n",
      "Requirement already satisfied: pillow>=8 in d:\\anaconda\\lib\\site-packages (from matplotlib) (10.3.0)\n",
      "Requirement already satisfied: pyparsing>=2.3.1 in d:\\anaconda\\lib\\site-packages (from matplotlib) (3.0.9)\n",
      "Requirement already satisfied: python-dateutil>=2.7 in c:\\users\\harry\\appdata\\roaming\\python\\python312\\site-packages (from matplotlib) (2.9.0.post0)\n",
      "Collecting numpy>=1.21 (from matplotlib)\n",
      "  Using cached numpy-1.26.4-cp312-cp312-win_amd64.whl.metadata (61 kB)\n",
      "Requirement already satisfied: six>=1.5 in c:\\users\\harry\\appdata\\roaming\\python\\python312\\site-packages (from python-dateutil>=2.7->matplotlib) (1.16.0)\n",
      "Using cached numpy-1.26.4-cp312-cp312-win_amd64.whl (15.5 MB)\n",
      "Installing collected packages: numpy\n",
      "  Attempting uninstall: numpy\n",
      "    Found existing installation: numpy 2.1.0\n",
      "    Uninstalling numpy-2.1.0:\n",
      "      Successfully uninstalled numpy-2.1.0\n",
      "Successfully installed numpy-1.26.4\n",
      "Note: you may need to restart the kernel to use updated packages.\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "ERROR: pip's dependency resolver does not currently take into account all the packages that are installed. This behaviour is the source of the following dependency conflicts.\n",
      "streamlit 1.32.0 requires packaging<24,>=16.8, but you have packaging 24.1 which is incompatible.\n"
     ]
    }
   ],
   "source": [
    "pip install matplotlib"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Requirement already satisfied: numpy in d:\\anaconda\\lib\\site-packages (1.26.4)\n",
      "Collecting numpy\n",
      "  Using cached numpy-2.1.0-cp312-cp312-win_amd64.whl.metadata (59 kB)\n",
      "Using cached numpy-2.1.0-cp312-cp312-win_amd64.whl (12.6 MB)\n",
      "Installing collected packages: numpy\n",
      "  Attempting uninstall: numpy\n",
      "    Found existing installation: numpy 1.26.4\n",
      "    Uninstalling numpy-1.26.4:\n",
      "      Successfully uninstalled numpy-1.26.4\n",
      "Successfully installed numpy-2.1.0\n",
      "Note: you may need to restart the kernel to use updated packages.\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "  WARNING: Failed to remove contents in a temporary directory 'D:\\anaconda\\Lib\\site-packages\\~=mpy.libs'.\n",
      "  You can safely remove it manually.\n",
      "  WARNING: Failed to remove contents in a temporary directory 'D:\\anaconda\\Lib\\site-packages\\~=mpy'.\n",
      "  You can safely remove it manually.\n",
      "ERROR: pip's dependency resolver does not currently take into account all the packages that are installed. This behaviour is the source of the following dependency conflicts.\n",
      "contourpy 1.2.0 requires numpy<2.0,>=1.20, but you have numpy 2.1.0 which is incompatible.\n",
      "numba 0.59.1 requires numpy<1.27,>=1.22, but you have numpy 2.1.0 which is incompatible.\n",
      "pywavelets 1.5.0 requires numpy<2.0,>=1.22.4, but you have numpy 2.1.0 which is incompatible.\n",
      "streamlit 1.32.0 requires numpy<2,>=1.19.3, but you have numpy 2.1.0 which is incompatible.\n",
      "streamlit 1.32.0 requires packaging<24,>=16.8, but you have packaging 24.1 which is incompatible.\n"
     ]
    }
   ],
   "source": [
    "pip install --upgrade numpy\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "[autoreload of numpy failed: Traceback (most recent call last):\n",
      "  File \"C:\\Users\\Harry\\AppData\\Roaming\\Python\\Python312\\site-packages\\IPython\\extensions\\autoreload.py\", line 276, in check\n",
      "    superreload(m, reload, self.old_objects)\n",
      "  File \"C:\\Users\\Harry\\AppData\\Roaming\\Python\\Python312\\site-packages\\IPython\\extensions\\autoreload.py\", line 475, in superreload\n",
      "    module = reload(module)\n",
      "             ^^^^^^^^^^^^^^\n",
      "  File \"D:\\anaconda\\Lib\\importlib\\__init__.py\", line 131, in reload\n",
      "    _bootstrap._exec(spec, module)\n",
      "  File \"<frozen importlib._bootstrap>\", line 866, in _exec\n",
      "  File \"<frozen importlib._bootstrap_external>\", line 995, in exec_module\n",
      "  File \"<frozen importlib._bootstrap>\", line 488, in _call_with_frames_removed\n",
      "  File \"D:\\anaconda\\Lib\\site-packages\\numpy\\__init__.py\", line 136, in <module>\n",
      "    from ._core import (\n",
      "  File \"D:\\anaconda\\Lib\\site-packages\\numpy\\_core\\memmap.py\", line 5, in <module>\n",
      "    from .numeric import uint8, ndarray, dtype\n",
      "  File \"D:\\anaconda\\Lib\\site-packages\\numpy\\_core\\numeric.py\", line 11, in <module>\n",
      "    from . import multiarray\n",
      "  File \"D:\\anaconda\\Lib\\site-packages\\numpy\\_core\\multiarray.py\", line 82, in <module>\n",
      "    @array_function_from_c_func_and_dispatcher(_multiarray_umath.empty_like)\n",
      "     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "RuntimeError: empty_like method already has a different docstring\n",
      "]\n",
      "[autoreload of numpy.core.multiarray failed: Traceback (most recent call last):\n",
      "  File \"C:\\Users\\Harry\\AppData\\Roaming\\Python\\Python312\\site-packages\\IPython\\extensions\\autoreload.py\", line 276, in check\n",
      "    superreload(m, reload, self.old_objects)\n",
      "  File \"C:\\Users\\Harry\\AppData\\Roaming\\Python\\Python312\\site-packages\\IPython\\extensions\\autoreload.py\", line 475, in superreload\n",
      "    module = reload(module)\n",
      "             ^^^^^^^^^^^^^^\n",
      "  File \"D:\\anaconda\\Lib\\importlib\\__init__.py\", line 131, in reload\n",
      "    _bootstrap._exec(spec, module)\n",
      "  File \"<frozen importlib._bootstrap>\", line 866, in _exec\n",
      "  File \"<frozen importlib._bootstrap_external>\", line 995, in exec_module\n",
      "  File \"<frozen importlib._bootstrap>\", line 488, in _call_with_frames_removed\n",
      "  File \"D:\\anaconda\\Lib\\site-packages\\numpy\\core\\multiarray.py\", line 1, in <module>\n",
      "    from numpy._core import multiarray\n",
      "  File \"D:\\anaconda\\Lib\\site-packages\\numpy\\_core\\multiarray.py\", line 82, in <module>\n",
      "    @array_function_from_c_func_and_dispatcher(_multiarray_umath.empty_like)\n",
      "     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"D:\\anaconda\\Lib\\site-packages\\numpy\\_core\\overrides.py\", line 178, in decorator\n",
      "    return array_function_dispatch(\n",
      "           ^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"D:\\anaconda\\Lib\\site-packages\\numpy\\_core\\overrides.py\", line 158, in decorator\n",
      "    add_docstring(implementation, dispatcher.__doc__)\n",
      "RuntimeError: empty_like method already has a different docstring\n",
      "]\n",
      "[autoreload of numpy.core.fromnumeric failed: Traceback (most recent call last):\n",
      "  File \"C:\\Users\\Harry\\AppData\\Roaming\\Python\\Python312\\site-packages\\IPython\\extensions\\autoreload.py\", line 276, in check\n",
      "    superreload(m, reload, self.old_objects)\n",
      "  File \"C:\\Users\\Harry\\AppData\\Roaming\\Python\\Python312\\site-packages\\IPython\\extensions\\autoreload.py\", line 455, in superreload\n",
      "    if not append_obj(module, old_objects, name, obj):\n",
      "           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"C:\\Users\\Harry\\AppData\\Roaming\\Python\\Python312\\site-packages\\IPython\\extensions\\autoreload.py\", line 423, in append_obj\n",
      "    in_module = hasattr(obj, \"__module__\") and obj.__module__ == module.__name__\n",
      "                ^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"D:\\anaconda\\Lib\\site-packages\\numpy\\core\\umath.py\", line 2, in __getattr__\n",
      "    from numpy._core import umath\n",
      "  File \"D:\\anaconda\\Lib\\site-packages\\numpy\\_core\\umath.py\", line 15, in <module>\n",
      "    from ._multiarray_umath import (\n",
      "ImportError: cannot import name '_get_extobj_dict' from 'numpy.core._multiarray_umath' (D:\\anaconda\\Lib\\site-packages\\numpy\\core\\_multiarray_umath.cp312-win_amd64.pyd)\n",
      "]\n",
      "[autoreload of numpy.core.arrayprint failed: Traceback (most recent call last):\n",
      "  File \"C:\\Users\\Harry\\AppData\\Roaming\\Python\\Python312\\site-packages\\IPython\\extensions\\autoreload.py\", line 276, in check\n",
      "    superreload(m, reload, self.old_objects)\n",
      "  File \"C:\\Users\\Harry\\AppData\\Roaming\\Python\\Python312\\site-packages\\IPython\\extensions\\autoreload.py\", line 455, in superreload\n",
      "    if not append_obj(module, old_objects, name, obj):\n",
      "           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"C:\\Users\\Harry\\AppData\\Roaming\\Python\\Python312\\site-packages\\IPython\\extensions\\autoreload.py\", line 423, in append_obj\n",
      "    in_module = hasattr(obj, \"__module__\") and obj.__module__ == module.__name__\n",
      "                ^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"D:\\anaconda\\Lib\\site-packages\\numpy\\core\\numerictypes.py\", line 2, in __getattr__\n",
      "    from numpy._core import numerictypes\n",
      "  File \"D:\\anaconda\\Lib\\site-packages\\numpy\\_core\\numerictypes.py\", line 82, in <module>\n",
      "    from . import multiarray as ma\n",
      "  File \"D:\\anaconda\\Lib\\site-packages\\numpy\\_core\\multiarray.py\", line 82, in <module>\n",
      "    @array_function_from_c_func_and_dispatcher(_multiarray_umath.empty_like)\n",
      "     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"D:\\anaconda\\Lib\\site-packages\\numpy\\_core\\overrides.py\", line 178, in decorator\n",
      "    return array_function_dispatch(\n",
      "           ^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"D:\\anaconda\\Lib\\site-packages\\numpy\\_core\\overrides.py\", line 158, in decorator\n",
      "    add_docstring(implementation, dispatcher.__doc__)\n",
      "RuntimeError: empty_like method already has a different docstring\n",
      "]\n",
      "[autoreload of numpy.core.numeric failed: Traceback (most recent call last):\n",
      "  File \"C:\\Users\\Harry\\AppData\\Roaming\\Python\\Python312\\site-packages\\IPython\\extensions\\autoreload.py\", line 276, in check\n",
      "    superreload(m, reload, self.old_objects)\n",
      "  File \"C:\\Users\\Harry\\AppData\\Roaming\\Python\\Python312\\site-packages\\IPython\\extensions\\autoreload.py\", line 455, in superreload\n",
      "    if not append_obj(module, old_objects, name, obj):\n",
      "           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"C:\\Users\\Harry\\AppData\\Roaming\\Python\\Python312\\site-packages\\IPython\\extensions\\autoreload.py\", line 423, in append_obj\n",
      "    in_module = hasattr(obj, \"__module__\") and obj.__module__ == module.__name__\n",
      "                ^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"D:\\anaconda\\Lib\\site-packages\\numpy\\core\\umath.py\", line 2, in __getattr__\n",
      "    from numpy._core import umath\n",
      "  File \"D:\\anaconda\\Lib\\site-packages\\numpy\\_core\\umath.py\", line 15, in <module>\n",
      "    from ._multiarray_umath import (\n",
      "ImportError: cannot import name '_get_extobj_dict' from 'numpy.core._multiarray_umath' (D:\\anaconda\\Lib\\site-packages\\numpy\\core\\_multiarray_umath.cp312-win_amd64.pyd)\n",
      "]\n",
      "[autoreload of numpy.core.records failed: Traceback (most recent call last):\n",
      "  File \"C:\\Users\\Harry\\AppData\\Roaming\\Python\\Python312\\site-packages\\IPython\\extensions\\autoreload.py\", line 276, in check\n",
      "    superreload(m, reload, self.old_objects)\n",
      "  File \"C:\\Users\\Harry\\AppData\\Roaming\\Python\\Python312\\site-packages\\IPython\\extensions\\autoreload.py\", line 455, in superreload\n",
      "    if not append_obj(module, old_objects, name, obj):\n",
      "           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"C:\\Users\\Harry\\AppData\\Roaming\\Python\\Python312\\site-packages\\IPython\\extensions\\autoreload.py\", line 423, in append_obj\n",
      "    in_module = hasattr(obj, \"__module__\") and obj.__module__ == module.__name__\n",
      "                ^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"D:\\anaconda\\Lib\\site-packages\\numpy\\core\\numerictypes.py\", line 2, in __getattr__\n",
      "    from numpy._core import numerictypes\n",
      "  File \"D:\\anaconda\\Lib\\site-packages\\numpy\\_core\\numerictypes.py\", line 82, in <module>\n",
      "    from . import multiarray as ma\n",
      "  File \"D:\\anaconda\\Lib\\site-packages\\numpy\\_core\\multiarray.py\", line 82, in <module>\n",
      "    @array_function_from_c_func_and_dispatcher(_multiarray_umath.empty_like)\n",
      "     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"D:\\anaconda\\Lib\\site-packages\\numpy\\_core\\overrides.py\", line 178, in decorator\n",
      "    return array_function_dispatch(\n",
      "           ^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"D:\\anaconda\\Lib\\site-packages\\numpy\\_core\\overrides.py\", line 158, in decorator\n",
      "    add_docstring(implementation, dispatcher.__doc__)\n",
      "RuntimeError: empty_like method already has a different docstring\n",
      "]\n",
      "[autoreload of numpy.core.getlimits failed: Traceback (most recent call last):\n",
      "  File \"C:\\Users\\Harry\\AppData\\Roaming\\Python\\Python312\\site-packages\\IPython\\extensions\\autoreload.py\", line 276, in check\n",
      "    superreload(m, reload, self.old_objects)\n",
      "  File \"C:\\Users\\Harry\\AppData\\Roaming\\Python\\Python312\\site-packages\\IPython\\extensions\\autoreload.py\", line 455, in superreload\n",
      "    if not append_obj(module, old_objects, name, obj):\n",
      "           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"C:\\Users\\Harry\\AppData\\Roaming\\Python\\Python312\\site-packages\\IPython\\extensions\\autoreload.py\", line 423, in append_obj\n",
      "    in_module = hasattr(obj, \"__module__\") and obj.__module__ == module.__name__\n",
      "                ^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"D:\\anaconda\\Lib\\site-packages\\numpy\\core\\numerictypes.py\", line 2, in __getattr__\n",
      "    from numpy._core import numerictypes\n",
      "  File \"D:\\anaconda\\Lib\\site-packages\\numpy\\_core\\numerictypes.py\", line 82, in <module>\n",
      "    from . import multiarray as ma\n",
      "  File \"D:\\anaconda\\Lib\\site-packages\\numpy\\_core\\multiarray.py\", line 82, in <module>\n",
      "    @array_function_from_c_func_and_dispatcher(_multiarray_umath.empty_like)\n",
      "     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"D:\\anaconda\\Lib\\site-packages\\numpy\\_core\\overrides.py\", line 178, in decorator\n",
      "    return array_function_dispatch(\n",
      "           ^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"D:\\anaconda\\Lib\\site-packages\\numpy\\_core\\overrides.py\", line 158, in decorator\n",
      "    add_docstring(implementation, dispatcher.__doc__)\n",
      "RuntimeError: empty_like method already has a different docstring\n",
      "]\n",
      "[autoreload of numpy.core._internal failed: Traceback (most recent call last):\n",
      "  File \"C:\\Users\\Harry\\AppData\\Roaming\\Python\\Python312\\site-packages\\IPython\\extensions\\autoreload.py\", line 276, in check\n",
      "    superreload(m, reload, self.old_objects)\n",
      "  File \"C:\\Users\\Harry\\AppData\\Roaming\\Python\\Python312\\site-packages\\IPython\\extensions\\autoreload.py\", line 475, in superreload\n",
      "    module = reload(module)\n",
      "             ^^^^^^^^^^^^^^\n",
      "  File \"D:\\anaconda\\Lib\\importlib\\__init__.py\", line 131, in reload\n",
      "    _bootstrap._exec(spec, module)\n",
      "  File \"<frozen importlib._bootstrap>\", line 866, in _exec\n",
      "  File \"<frozen importlib._bootstrap_external>\", line 995, in exec_module\n",
      "  File \"<frozen importlib._bootstrap>\", line 488, in _call_with_frames_removed\n",
      "  File \"D:\\anaconda\\Lib\\site-packages\\numpy\\core\\_internal.py\", line 1, in <module>\n",
      "    from numpy._core import _internal\n",
      "  File \"D:\\anaconda\\Lib\\site-packages\\numpy\\_core\\_internal.py\", line 14, in <module>\n",
      "    from .multiarray import dtype, array, ndarray, promote_types, StringDType\n",
      "  File \"D:\\anaconda\\Lib\\site-packages\\numpy\\_core\\multiarray.py\", line 82, in <module>\n",
      "    @array_function_from_c_func_and_dispatcher(_multiarray_umath.empty_like)\n",
      "     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"D:\\anaconda\\Lib\\site-packages\\numpy\\_core\\overrides.py\", line 178, in decorator\n",
      "    return array_function_dispatch(\n",
      "           ^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "RuntimeError: empty_like method already has a different docstring\n",
      "]\n",
      "[autoreload of numpy.core failed: Traceback (most recent call last):\n",
      "  File \"C:\\Users\\Harry\\AppData\\Roaming\\Python\\Python312\\site-packages\\IPython\\extensions\\autoreload.py\", line 276, in check\n",
      "    superreload(m, reload, self.old_objects)\n",
      "  File \"C:\\Users\\Harry\\AppData\\Roaming\\Python\\Python312\\site-packages\\IPython\\extensions\\autoreload.py\", line 455, in superreload\n",
      "    if not append_obj(module, old_objects, name, obj):\n",
      "           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"C:\\Users\\Harry\\AppData\\Roaming\\Python\\Python312\\site-packages\\IPython\\extensions\\autoreload.py\", line 423, in append_obj\n",
      "    in_module = hasattr(obj, \"__module__\") and obj.__module__ == module.__name__\n",
      "                ^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"D:\\anaconda\\Lib\\site-packages\\numpy\\core\\umath.py\", line 2, in __getattr__\n",
      "    from numpy._core import umath\n",
      "  File \"D:\\anaconda\\Lib\\site-packages\\numpy\\_core\\umath.py\", line 15, in <module>\n",
      "    from ._multiarray_umath import (\n",
      "ImportError: cannot import name '_get_extobj_dict' from 'numpy.core._multiarray_umath' (D:\\anaconda\\Lib\\site-packages\\numpy\\core\\_multiarray_umath.cp312-win_amd64.pyd)\n",
      "]\n",
      "[autoreload of numpy.lib.mixins failed: Traceback (most recent call last):\n",
      "  File \"C:\\Users\\Harry\\AppData\\Roaming\\Python\\Python312\\site-packages\\IPython\\extensions\\autoreload.py\", line 276, in check\n",
      "    superreload(m, reload, self.old_objects)\n",
      "  File \"C:\\Users\\Harry\\AppData\\Roaming\\Python\\Python312\\site-packages\\IPython\\extensions\\autoreload.py\", line 455, in superreload\n",
      "    if not append_obj(module, old_objects, name, obj):\n",
      "           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"C:\\Users\\Harry\\AppData\\Roaming\\Python\\Python312\\site-packages\\IPython\\extensions\\autoreload.py\", line 423, in append_obj\n",
      "    in_module = hasattr(obj, \"__module__\") and obj.__module__ == module.__name__\n",
      "                ^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"D:\\anaconda\\Lib\\site-packages\\numpy\\core\\umath.py\", line 2, in __getattr__\n",
      "    from numpy._core import umath\n",
      "  File \"D:\\anaconda\\Lib\\site-packages\\numpy\\_core\\umath.py\", line 15, in <module>\n",
      "    from ._multiarray_umath import (\n",
      "ImportError: cannot import name '_get_extobj_dict' from 'numpy.core._multiarray_umath' (D:\\anaconda\\Lib\\site-packages\\numpy\\core\\_multiarray_umath.cp312-win_amd64.pyd)\n",
      "]\n",
      "[autoreload of numpy.lib.scimath failed: Traceback (most recent call last):\n",
      "  File \"C:\\Users\\Harry\\AppData\\Roaming\\Python\\Python312\\site-packages\\IPython\\extensions\\autoreload.py\", line 276, in check\n",
      "    superreload(m, reload, self.old_objects)\n",
      "  File \"C:\\Users\\Harry\\AppData\\Roaming\\Python\\Python312\\site-packages\\IPython\\extensions\\autoreload.py\", line 455, in superreload\n",
      "    if not append_obj(module, old_objects, name, obj):\n",
      "           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"C:\\Users\\Harry\\AppData\\Roaming\\Python\\Python312\\site-packages\\IPython\\extensions\\autoreload.py\", line 423, in append_obj\n",
      "    in_module = hasattr(obj, \"__module__\") and obj.__module__ == module.__name__\n",
      "                ^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"D:\\anaconda\\Lib\\site-packages\\numpy\\core\\numerictypes.py\", line 2, in __getattr__\n",
      "    from numpy._core import numerictypes\n",
      "  File \"D:\\anaconda\\Lib\\site-packages\\numpy\\_core\\numerictypes.py\", line 82, in <module>\n",
      "    from . import multiarray as ma\n",
      "  File \"D:\\anaconda\\Lib\\site-packages\\numpy\\_core\\multiarray.py\", line 82, in <module>\n",
      "    @array_function_from_c_func_and_dispatcher(_multiarray_umath.empty_like)\n",
      "     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"D:\\anaconda\\Lib\\site-packages\\numpy\\_core\\overrides.py\", line 178, in decorator\n",
      "    return array_function_dispatch(\n",
      "           ^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"D:\\anaconda\\Lib\\site-packages\\numpy\\_core\\overrides.py\", line 158, in decorator\n",
      "    add_docstring(implementation, dispatcher.__doc__)\n",
      "RuntimeError: empty_like method already has a different docstring\n",
      "]\n",
      "[autoreload of numpy.lib.stride_tricks failed: Traceback (most recent call last):\n",
      "  File \"C:\\Users\\Harry\\AppData\\Roaming\\Python\\Python312\\site-packages\\IPython\\extensions\\autoreload.py\", line 276, in check\n",
      "    superreload(m, reload, self.old_objects)\n",
      "  File \"C:\\Users\\Harry\\AppData\\Roaming\\Python\\Python312\\site-packages\\IPython\\extensions\\autoreload.py\", line 475, in superreload\n",
      "    module = reload(module)\n",
      "             ^^^^^^^^^^^^^^\n",
      "  File \"D:\\anaconda\\Lib\\importlib\\__init__.py\", line 131, in reload\n",
      "    _bootstrap._exec(spec, module)\n",
      "  File \"<frozen importlib._bootstrap>\", line 866, in _exec\n",
      "  File \"<frozen importlib._bootstrap_external>\", line 995, in exec_module\n",
      "  File \"<frozen importlib._bootstrap>\", line 488, in _call_with_frames_removed\n",
      "  File \"D:\\anaconda\\Lib\\site-packages\\numpy\\lib\\stride_tricks.py\", line 1, in <module>\n",
      "    from ._stride_tricks_impl import (\n",
      "  File \"D:\\anaconda\\Lib\\site-packages\\numpy\\lib\\_stride_tricks_impl.py\", line 14, in <module>\n",
      "    from numpy._core.numeric import normalize_axis_tuple\n",
      "  File \"D:\\anaconda\\Lib\\site-packages\\numpy\\_core\\numeric.py\", line 11, in <module>\n",
      "    from . import multiarray\n",
      "  File \"D:\\anaconda\\Lib\\site-packages\\numpy\\_core\\multiarray.py\", line 82, in <module>\n",
      "    @array_function_from_c_func_and_dispatcher(_multiarray_umath.empty_like)\n",
      "     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "RuntimeError: empty_like method already has a different docstring\n",
      "]\n",
      "D:\\anaconda\\Lib\\site-packages\\numpy\\_typing\\_scalars.py:12: FutureWarning: In the future `np.bool` will be defined as the corresponding NumPy scalar.\n",
      "  _BoolLike_co: TypeAlias = bool | np.bool\n",
      "[autoreload of numpy._typing._scalars failed: Traceback (most recent call last):\n",
      "  File \"C:\\Users\\Harry\\AppData\\Roaming\\Python\\Python312\\site-packages\\IPython\\extensions\\autoreload.py\", line 276, in check\n",
      "    superreload(m, reload, self.old_objects)\n",
      "  File \"C:\\Users\\Harry\\AppData\\Roaming\\Python\\Python312\\site-packages\\IPython\\extensions\\autoreload.py\", line 475, in superreload\n",
      "    module = reload(module)\n",
      "             ^^^^^^^^^^^^^^\n",
      "  File \"D:\\anaconda\\Lib\\importlib\\__init__.py\", line 131, in reload\n",
      "    _bootstrap._exec(spec, module)\n",
      "  File \"<frozen importlib._bootstrap>\", line 866, in _exec\n",
      "  File \"<frozen importlib._bootstrap_external>\", line 995, in exec_module\n",
      "  File \"<frozen importlib._bootstrap>\", line 488, in _call_with_frames_removed\n",
      "  File \"D:\\anaconda\\Lib\\site-packages\\numpy\\_typing\\_scalars.py\", line 12, in <module>\n",
      "    _BoolLike_co: TypeAlias = bool | np.bool\n",
      "                                     ^^^^^^^\n",
      "  File \"D:\\anaconda\\Lib\\site-packages\\numpy\\__init__.py\", line 338, in __getattr__\n",
      "    warnings.filterwarnings(\"ignore\", message=\"numpy.ndarray size changed\")\n",
      "            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "AttributeError: module 'numpy' has no attribute 'bool'.\n",
      "`np.bool` was a deprecated alias for the builtin `bool`. To avoid this error in existing code, use `bool` by itself. Doing this will not modify any behavior and is safe. If you specifically wanted the numpy scalar type, use `np.bool_` here.\n",
      "The aliases was originally deprecated in NumPy 1.20; for more details and guidance see the original release note at:\n",
      "    https://numpy.org/devdocs/release/1.20.0-notes.html#deprecations. Did you mean: 'bool_'?\n",
      "]\n",
      "D:\\anaconda\\Lib\\site-packages\\numpy\\_typing\\_dtype_like.py:141: FutureWarning: In the future `np.bool` will be defined as the corresponding NumPy scalar.\n",
      "  | type[np.bool]\n",
      "[autoreload of numpy._typing._dtype_like failed: Traceback (most recent call last):\n",
      "  File \"C:\\Users\\Harry\\AppData\\Roaming\\Python\\Python312\\site-packages\\IPython\\extensions\\autoreload.py\", line 276, in check\n",
      "    superreload(m, reload, self.old_objects)\n",
      "  File \"C:\\Users\\Harry\\AppData\\Roaming\\Python\\Python312\\site-packages\\IPython\\extensions\\autoreload.py\", line 475, in superreload\n",
      "    module = reload(module)\n",
      "             ^^^^^^^^^^^^^^\n",
      "  File \"D:\\anaconda\\Lib\\importlib\\__init__.py\", line 131, in reload\n",
      "    _bootstrap._exec(spec, module)\n",
      "  File \"<frozen importlib._bootstrap>\", line 866, in _exec\n",
      "  File \"<frozen importlib._bootstrap_external>\", line 995, in exec_module\n",
      "  File \"<frozen importlib._bootstrap>\", line 488, in _call_with_frames_removed\n",
      "  File \"D:\\anaconda\\Lib\\site-packages\\numpy\\_typing\\_dtype_like.py\", line 141, in <module>\n",
      "    | type[np.bool]\n",
      "           ^^^^^^^\n",
      "  File \"D:\\anaconda\\Lib\\site-packages\\numpy\\__init__.py\", line 338, in __getattr__\n",
      "    warnings.filterwarnings(\"ignore\", message=\"numpy.ndarray size changed\")\n",
      "            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "AttributeError: module 'numpy' has no attribute 'bool'.\n",
      "`np.bool` was a deprecated alias for the builtin `bool`. To avoid this error in existing code, use `bool` by itself. Doing this will not modify any behavior and is safe. If you specifically wanted the numpy scalar type, use `np.bool_` here.\n",
      "The aliases was originally deprecated in NumPy 1.20; for more details and guidance see the original release note at:\n",
      "    https://numpy.org/devdocs/release/1.20.0-notes.html#deprecations. Did you mean: 'bool_'?\n",
      "]\n",
      "D:\\anaconda\\Lib\\site-packages\\numpy\\_typing\\_array_like.py:97: FutureWarning: In the future `np.bool` will be defined as the corresponding NumPy scalar.\n",
      "  dtype[np.bool],\n",
      "[autoreload of numpy._typing._array_like failed: Traceback (most recent call last):\n",
      "  File \"C:\\Users\\Harry\\AppData\\Roaming\\Python\\Python312\\site-packages\\IPython\\extensions\\autoreload.py\", line 276, in check\n",
      "    superreload(m, reload, self.old_objects)\n",
      "  File \"C:\\Users\\Harry\\AppData\\Roaming\\Python\\Python312\\site-packages\\IPython\\extensions\\autoreload.py\", line 475, in superreload\n",
      "    module = reload(module)\n",
      "             ^^^^^^^^^^^^^^\n",
      "  File \"D:\\anaconda\\Lib\\importlib\\__init__.py\", line 131, in reload\n",
      "    _bootstrap._exec(spec, module)\n",
      "  File \"<frozen importlib._bootstrap>\", line 866, in _exec\n",
      "  File \"<frozen importlib._bootstrap_external>\", line 995, in exec_module\n",
      "  File \"<frozen importlib._bootstrap>\", line 488, in _call_with_frames_removed\n",
      "  File \"D:\\anaconda\\Lib\\site-packages\\numpy\\_typing\\_array_like.py\", line 97, in <module>\n",
      "    dtype[np.bool],\n",
      "          ^^^^^^^\n",
      "  File \"D:\\anaconda\\Lib\\site-packages\\numpy\\__init__.py\", line 338, in __getattr__\n",
      "    warnings.filterwarnings(\"ignore\", message=\"numpy.ndarray size changed\")\n",
      "            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "AttributeError: module 'numpy' has no attribute 'bool'.\n",
      "`np.bool` was a deprecated alias for the builtin `bool`. To avoid this error in existing code, use `bool` by itself. Doing this will not modify any behavior and is safe. If you specifically wanted the numpy scalar type, use `np.bool_` here.\n",
      "The aliases was originally deprecated in NumPy 1.20; for more details and guidance see the original release note at:\n",
      "    https://numpy.org/devdocs/release/1.20.0-notes.html#deprecations. Did you mean: 'bool_'?\n",
      "]\n",
      "[autoreload of numpy.linalg failed: Traceback (most recent call last):\n",
      "  File \"C:\\Users\\Harry\\AppData\\Roaming\\Python\\Python312\\site-packages\\IPython\\extensions\\autoreload.py\", line 276, in check\n",
      "    superreload(m, reload, self.old_objects)\n",
      "  File \"C:\\Users\\Harry\\AppData\\Roaming\\Python\\Python312\\site-packages\\IPython\\extensions\\autoreload.py\", line 455, in superreload\n",
      "    if not append_obj(module, old_objects, name, obj):\n",
      "           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"C:\\Users\\Harry\\AppData\\Roaming\\Python\\Python312\\site-packages\\IPython\\extensions\\autoreload.py\", line 423, in append_obj\n",
      "    in_module = hasattr(obj, \"__module__\") and obj.__module__ == module.__name__\n",
      "                ^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"D:\\anaconda\\Lib\\site-packages\\numpy\\linalg\\linalg.py\", line 3, in __getattr__\n",
      "    from numpy.linalg import _linalg\n",
      "  File \"D:\\anaconda\\Lib\\site-packages\\numpy\\linalg\\_linalg.py\", line 25, in <module>\n",
      "    from numpy._core import (\n",
      "ImportError: cannot import name 'array' from 'numpy._core' (D:\\anaconda\\Lib\\site-packages\\numpy\\_core\\__init__.py). Did you mean: 'asarray'?\n",
      "]\n",
      "[autoreload of numpy.matrixlib.defmatrix failed: Traceback (most recent call last):\n",
      "  File \"C:\\Users\\Harry\\AppData\\Roaming\\Python\\Python312\\site-packages\\IPython\\extensions\\autoreload.py\", line 276, in check\n",
      "    superreload(m, reload, self.old_objects)\n",
      "  File \"C:\\Users\\Harry\\AppData\\Roaming\\Python\\Python312\\site-packages\\IPython\\extensions\\autoreload.py\", line 475, in superreload\n",
      "    module = reload(module)\n",
      "             ^^^^^^^^^^^^^^\n",
      "  File \"D:\\anaconda\\Lib\\importlib\\__init__.py\", line 131, in reload\n",
      "    _bootstrap._exec(spec, module)\n",
      "  File \"<frozen importlib._bootstrap>\", line 866, in _exec\n",
      "  File \"<frozen importlib._bootstrap_external>\", line 995, in exec_module\n",
      "  File \"<frozen importlib._bootstrap>\", line 488, in _call_with_frames_removed\n",
      "  File \"D:\\anaconda\\Lib\\site-packages\\numpy\\matrixlib\\defmatrix.py\", line 8, in <module>\n",
      "    import numpy._core.numeric as N\n",
      "  File \"D:\\anaconda\\Lib\\site-packages\\numpy\\_core\\numeric.py\", line 11, in <module>\n",
      "    from . import multiarray\n",
      "  File \"D:\\anaconda\\Lib\\site-packages\\numpy\\_core\\multiarray.py\", line 82, in <module>\n",
      "    @array_function_from_c_func_and_dispatcher(_multiarray_umath.empty_like)\n",
      "     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"D:\\anaconda\\Lib\\site-packages\\numpy\\_core\\overrides.py\", line 178, in decorator\n",
      "    return array_function_dispatch(\n",
      "           ^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "RuntimeError: empty_like method already has a different docstring\n",
      "]\n",
      "[autoreload of numpy.lib.format failed: Traceback (most recent call last):\n",
      "  File \"C:\\Users\\Harry\\AppData\\Roaming\\Python\\Python312\\site-packages\\IPython\\extensions\\autoreload.py\", line 276, in check\n",
      "    superreload(m, reload, self.old_objects)\n",
      "  File \"C:\\Users\\Harry\\AppData\\Roaming\\Python\\Python312\\site-packages\\IPython\\extensions\\autoreload.py\", line 475, in superreload\n",
      "    module = reload(module)\n",
      "             ^^^^^^^^^^^^^^\n",
      "  File \"D:\\anaconda\\Lib\\importlib\\__init__.py\", line 131, in reload\n",
      "    _bootstrap._exec(spec, module)\n",
      "  File \"<frozen importlib._bootstrap>\", line 866, in _exec\n",
      "  File \"<frozen importlib._bootstrap_external>\", line 995, in exec_module\n",
      "  File \"<frozen importlib._bootstrap>\", line 488, in _call_with_frames_removed\n",
      "  File \"D:\\anaconda\\Lib\\site-packages\\numpy\\lib\\format.py\", line 170, in <module>\n",
      "    from numpy.lib._utils_impl import drop_metadata\n",
      "  File \"D:\\anaconda\\Lib\\site-packages\\numpy\\lib\\_utils_impl.py\", line 10, in <module>\n",
      "    from numpy._core import ndarray\n",
      "ImportError: cannot import name 'ndarray' from 'numpy._core' (D:\\anaconda\\Lib\\site-packages\\numpy\\_core\\__init__.py). Did you mean: 'array'?\n",
      "]\n",
      "[autoreload of numpy.lib._iotools failed: Traceback (most recent call last):\n",
      "  File \"C:\\Users\\Harry\\AppData\\Roaming\\Python\\Python312\\site-packages\\IPython\\extensions\\autoreload.py\", line 276, in check\n",
      "    superreload(m, reload, self.old_objects)\n",
      "  File \"C:\\Users\\Harry\\AppData\\Roaming\\Python\\Python312\\site-packages\\IPython\\extensions\\autoreload.py\", line 475, in superreload\n",
      "    module = reload(module)\n",
      "             ^^^^^^^^^^^^^^\n",
      "  File \"D:\\anaconda\\Lib\\importlib\\__init__.py\", line 131, in reload\n",
      "    _bootstrap._exec(spec, module)\n",
      "  File \"<frozen importlib._bootstrap>\", line 866, in _exec\n",
      "  File \"<frozen importlib._bootstrap_external>\", line 995, in exec_module\n",
      "  File \"<frozen importlib._bootstrap>\", line 488, in _call_with_frames_removed\n",
      "  File \"D:\\anaconda\\Lib\\site-packages\\numpy\\lib\\_iotools.py\", line 7, in <module>\n",
      "    import numpy._core.numeric as nx\n",
      "  File \"D:\\anaconda\\Lib\\site-packages\\numpy\\_core\\numeric.py\", line 11, in <module>\n",
      "    from . import multiarray\n",
      "  File \"D:\\anaconda\\Lib\\site-packages\\numpy\\_core\\multiarray.py\", line 82, in <module>\n",
      "    @array_function_from_c_func_and_dispatcher(_multiarray_umath.empty_like)\n",
      "     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"D:\\anaconda\\Lib\\site-packages\\numpy\\_core\\overrides.py\", line 178, in decorator\n",
      "    return array_function_dispatch(\n",
      "           ^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "RuntimeError: empty_like method already has a different docstring\n",
      "]\n",
      "[autoreload of numpy.lib.npyio failed: Traceback (most recent call last):\n",
      "  File \"C:\\Users\\Harry\\AppData\\Roaming\\Python\\Python312\\site-packages\\IPython\\extensions\\autoreload.py\", line 276, in check\n",
      "    superreload(m, reload, self.old_objects)\n",
      "  File \"C:\\Users\\Harry\\AppData\\Roaming\\Python\\Python312\\site-packages\\IPython\\extensions\\autoreload.py\", line 475, in superreload\n",
      "    module = reload(module)\n",
      "             ^^^^^^^^^^^^^^\n",
      "  File \"D:\\anaconda\\Lib\\importlib\\__init__.py\", line 131, in reload\n",
      "    _bootstrap._exec(spec, module)\n",
      "  File \"<frozen importlib._bootstrap>\", line 866, in _exec\n",
      "  File \"<frozen importlib._bootstrap_external>\", line 995, in exec_module\n",
      "  File \"<frozen importlib._bootstrap>\", line 488, in _call_with_frames_removed\n",
      "  File \"D:\\anaconda\\Lib\\site-packages\\numpy\\lib\\npyio.py\", line 1, in <module>\n",
      "    from ._npyio_impl import (\n",
      "  File \"D:\\anaconda\\Lib\\site-packages\\numpy\\lib\\_npyio_impl.py\", line 20, in <module>\n",
      "    from numpy._core.multiarray import packbits, unpackbits\n",
      "  File \"D:\\anaconda\\Lib\\site-packages\\numpy\\_core\\multiarray.py\", line 82, in <module>\n",
      "    @array_function_from_c_func_and_dispatcher(_multiarray_umath.empty_like)\n",
      "     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"D:\\anaconda\\Lib\\site-packages\\numpy\\_core\\overrides.py\", line 178, in decorator\n",
      "    return array_function_dispatch(\n",
      "           ^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "RuntimeError: empty_like method already has a different docstring\n",
      "]\n",
      "[autoreload of numpy.lib failed: Traceback (most recent call last):\n",
      "  File \"C:\\Users\\Harry\\AppData\\Roaming\\Python\\Python312\\site-packages\\IPython\\extensions\\autoreload.py\", line 276, in check\n",
      "    superreload(m, reload, self.old_objects)\n",
      "  File \"C:\\Users\\Harry\\AppData\\Roaming\\Python\\Python312\\site-packages\\IPython\\extensions\\autoreload.py\", line 475, in superreload\n",
      "    module = reload(module)\n",
      "             ^^^^^^^^^^^^^^\n",
      "  File \"D:\\anaconda\\Lib\\importlib\\__init__.py\", line 131, in reload\n",
      "    _bootstrap._exec(spec, module)\n",
      "  File \"<frozen importlib._bootstrap>\", line 866, in _exec\n",
      "  File \"<frozen importlib._bootstrap_external>\", line 995, in exec_module\n",
      "  File \"<frozen importlib._bootstrap>\", line 488, in _call_with_frames_removed\n",
      "  File \"D:\\anaconda\\Lib\\site-packages\\numpy\\lib\\__init__.py\", line 13, in <module>\n",
      "    from . import array_utils\n",
      "  File \"D:\\anaconda\\Lib\\site-packages\\numpy\\lib\\array_utils.py\", line 1, in <module>\n",
      "    from ._array_utils_impl import (\n",
      "  File \"D:\\anaconda\\Lib\\site-packages\\numpy\\lib\\_array_utils_impl.py\", line 4, in <module>\n",
      "    from numpy._core import asarray\n",
      "ImportError: cannot import name 'asarray' from 'numpy._core' (D:\\anaconda\\Lib\\site-packages\\numpy\\_core\\__init__.py). Did you mean: 'asfarray'?\n",
      "]\n",
      "[autoreload of numpy.fft._pocketfft failed: Traceback (most recent call last):\n",
      "  File \"C:\\Users\\Harry\\AppData\\Roaming\\Python\\Python312\\site-packages\\IPython\\extensions\\autoreload.py\", line 276, in check\n",
      "    superreload(m, reload, self.old_objects)\n",
      "  File \"C:\\Users\\Harry\\AppData\\Roaming\\Python\\Python312\\site-packages\\IPython\\extensions\\autoreload.py\", line 475, in superreload\n",
      "    module = reload(module)\n",
      "             ^^^^^^^^^^^^^^\n",
      "  File \"D:\\anaconda\\Lib\\importlib\\__init__.py\", line 131, in reload\n",
      "    _bootstrap._exec(spec, module)\n",
      "  File \"<frozen importlib._bootstrap>\", line 866, in _exec\n",
      "  File \"<frozen importlib._bootstrap_external>\", line 995, in exec_module\n",
      "  File \"<frozen importlib._bootstrap>\", line 488, in _call_with_frames_removed\n",
      "  File \"D:\\anaconda\\Lib\\site-packages\\numpy\\fft\\_pocketfft.py\", line 36, in <module>\n",
      "    from numpy.lib.array_utils import normalize_axis_index\n",
      "  File \"D:\\anaconda\\Lib\\site-packages\\numpy\\lib\\array_utils.py\", line 1, in <module>\n",
      "    from ._array_utils_impl import (\n",
      "  File \"D:\\anaconda\\Lib\\site-packages\\numpy\\lib\\_array_utils_impl.py\", line 4, in <module>\n",
      "    from numpy._core import asarray\n",
      "ImportError: cannot import name 'asarray' from 'numpy._core' (D:\\anaconda\\Lib\\site-packages\\numpy\\_core\\__init__.py). Did you mean: 'asfarray'?\n",
      "]\n",
      "[autoreload of numpy.fft failed: Traceback (most recent call last):\n",
      "  File \"C:\\Users\\Harry\\AppData\\Roaming\\Python\\Python312\\site-packages\\IPython\\extensions\\autoreload.py\", line 276, in check\n",
      "    superreload(m, reload, self.old_objects)\n",
      "  File \"C:\\Users\\Harry\\AppData\\Roaming\\Python\\Python312\\site-packages\\IPython\\extensions\\autoreload.py\", line 455, in superreload\n",
      "    if not append_obj(module, old_objects, name, obj):\n",
      "           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"C:\\Users\\Harry\\AppData\\Roaming\\Python\\Python312\\site-packages\\IPython\\extensions\\autoreload.py\", line 423, in append_obj\n",
      "    in_module = hasattr(obj, \"__module__\") and obj.__module__ == module.__name__\n",
      "                ^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"D:\\anaconda\\Lib\\site-packages\\numpy\\fft\\helper.py\", line 3, in __getattr__\n",
      "    from numpy.fft import _helper\n",
      "  File \"D:\\anaconda\\Lib\\site-packages\\numpy\\fft\\_helper.py\", line 5, in <module>\n",
      "    from numpy._core import integer, empty, arange, asarray, roll\n",
      "ImportError: cannot import name 'integer' from 'numpy._core' (D:\\anaconda\\Lib\\site-packages\\numpy\\_core\\__init__.py)\n",
      "]\n",
      "[autoreload of numpy.polynomial.polyutils failed: Traceback (most recent call last):\n",
      "  File \"C:\\Users\\Harry\\AppData\\Roaming\\Python\\Python312\\site-packages\\IPython\\extensions\\autoreload.py\", line 276, in check\n",
      "    superreload(m, reload, self.old_objects)\n",
      "  File \"C:\\Users\\Harry\\AppData\\Roaming\\Python\\Python312\\site-packages\\IPython\\extensions\\autoreload.py\", line 475, in superreload\n",
      "    module = reload(module)\n",
      "             ^^^^^^^^^^^^^^\n",
      "  File \"D:\\anaconda\\Lib\\importlib\\__init__.py\", line 131, in reload\n",
      "    _bootstrap._exec(spec, module)\n",
      "  File \"<frozen importlib._bootstrap>\", line 866, in _exec\n",
      "  File \"<frozen importlib._bootstrap_external>\", line 995, in exec_module\n",
      "  File \"<frozen importlib._bootstrap>\", line 488, in _call_with_frames_removed\n",
      "  File \"D:\\anaconda\\Lib\\site-packages\\numpy\\polynomial\\polyutils.py\", line 27, in <module>\n",
      "    from numpy._core.multiarray import dragon4_positional, dragon4_scientific\n",
      "  File \"D:\\anaconda\\Lib\\site-packages\\numpy\\_core\\multiarray.py\", line 82, in <module>\n",
      "    @array_function_from_c_func_and_dispatcher(_multiarray_umath.empty_like)\n",
      "     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"D:\\anaconda\\Lib\\site-packages\\numpy\\_core\\overrides.py\", line 178, in decorator\n",
      "    return array_function_dispatch(\n",
      "           ^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"D:\\anaconda\\Lib\\site-packages\\numpy\\_core\\overrides.py\", line 158, in decorator\n",
      "    add_docstring(implementation, dispatcher.__doc__)\n",
      "RuntimeError: empty_like method already has a different docstring\n",
      "]\n",
      "[autoreload of numpy.polynomial.polynomial failed: Traceback (most recent call last):\n",
      "  File \"C:\\Users\\Harry\\AppData\\Roaming\\Python\\Python312\\site-packages\\IPython\\extensions\\autoreload.py\", line 276, in check\n",
      "    superreload(m, reload, self.old_objects)\n",
      "  File \"C:\\Users\\Harry\\AppData\\Roaming\\Python\\Python312\\site-packages\\IPython\\extensions\\autoreload.py\", line 475, in superreload\n",
      "    module = reload(module)\n",
      "             ^^^^^^^^^^^^^^\n",
      "  File \"D:\\anaconda\\Lib\\importlib\\__init__.py\", line 131, in reload\n",
      "    _bootstrap._exec(spec, module)\n",
      "  File \"<frozen importlib._bootstrap>\", line 866, in _exec\n",
      "  File \"<frozen importlib._bootstrap_external>\", line 995, in exec_module\n",
      "  File \"<frozen importlib._bootstrap>\", line 488, in _call_with_frames_removed\n",
      "  File \"D:\\anaconda\\Lib\\site-packages\\numpy\\polynomial\\polynomial.py\", line 85, in <module>\n",
      "    from numpy.lib.array_utils import normalize_axis_index\n",
      "  File \"D:\\anaconda\\Lib\\site-packages\\numpy\\lib\\array_utils.py\", line 1, in <module>\n",
      "    from ._array_utils_impl import (\n",
      "  File \"D:\\anaconda\\Lib\\site-packages\\numpy\\lib\\_array_utils_impl.py\", line 4, in <module>\n",
      "    from numpy._core import asarray\n",
      "ImportError: cannot import name 'asarray' from 'numpy._core' (D:\\anaconda\\Lib\\site-packages\\numpy\\_core\\__init__.py). Did you mean: 'asfarray'?\n",
      "]\n",
      "[autoreload of numpy.polynomial.chebyshev failed: Traceback (most recent call last):\n",
      "  File \"C:\\Users\\Harry\\AppData\\Roaming\\Python\\Python312\\site-packages\\IPython\\extensions\\autoreload.py\", line 276, in check\n",
      "    superreload(m, reload, self.old_objects)\n",
      "  File \"C:\\Users\\Harry\\AppData\\Roaming\\Python\\Python312\\site-packages\\IPython\\extensions\\autoreload.py\", line 475, in superreload\n",
      "    module = reload(module)\n",
      "             ^^^^^^^^^^^^^^\n",
      "  File \"D:\\anaconda\\Lib\\importlib\\__init__.py\", line 131, in reload\n",
      "    _bootstrap._exec(spec, module)\n",
      "  File \"<frozen importlib._bootstrap>\", line 866, in _exec\n",
      "  File \"<frozen importlib._bootstrap_external>\", line 995, in exec_module\n",
      "  File \"<frozen importlib._bootstrap>\", line 488, in _call_with_frames_removed\n",
      "  File \"D:\\anaconda\\Lib\\site-packages\\numpy\\polynomial\\chebyshev.py\", line 112, in <module>\n",
      "    from numpy.lib.array_utils import normalize_axis_index\n",
      "  File \"D:\\anaconda\\Lib\\site-packages\\numpy\\lib\\array_utils.py\", line 1, in <module>\n",
      "    from ._array_utils_impl import (\n",
      "  File \"D:\\anaconda\\Lib\\site-packages\\numpy\\lib\\_array_utils_impl.py\", line 4, in <module>\n",
      "    from numpy._core import asarray\n",
      "ImportError: cannot import name 'asarray' from 'numpy._core' (D:\\anaconda\\Lib\\site-packages\\numpy\\_core\\__init__.py). Did you mean: 'asfarray'?\n",
      "]\n",
      "[autoreload of numpy.polynomial.legendre failed: Traceback (most recent call last):\n",
      "  File \"C:\\Users\\Harry\\AppData\\Roaming\\Python\\Python312\\site-packages\\IPython\\extensions\\autoreload.py\", line 276, in check\n",
      "    superreload(m, reload, self.old_objects)\n",
      "  File \"C:\\Users\\Harry\\AppData\\Roaming\\Python\\Python312\\site-packages\\IPython\\extensions\\autoreload.py\", line 475, in superreload\n",
      "    module = reload(module)\n",
      "             ^^^^^^^^^^^^^^\n",
      "  File \"D:\\anaconda\\Lib\\importlib\\__init__.py\", line 131, in reload\n",
      "    _bootstrap._exec(spec, module)\n",
      "  File \"<frozen importlib._bootstrap>\", line 866, in _exec\n",
      "  File \"<frozen importlib._bootstrap_external>\", line 995, in exec_module\n",
      "  File \"<frozen importlib._bootstrap>\", line 488, in _call_with_frames_removed\n",
      "  File \"D:\\anaconda\\Lib\\site-packages\\numpy\\polynomial\\legendre.py\", line 84, in <module>\n",
      "    from numpy.lib.array_utils import normalize_axis_index\n",
      "  File \"D:\\anaconda\\Lib\\site-packages\\numpy\\lib\\array_utils.py\", line 1, in <module>\n",
      "    from ._array_utils_impl import (\n",
      "  File \"D:\\anaconda\\Lib\\site-packages\\numpy\\lib\\_array_utils_impl.py\", line 4, in <module>\n",
      "    from numpy._core import asarray\n",
      "ImportError: cannot import name 'asarray' from 'numpy._core' (D:\\anaconda\\Lib\\site-packages\\numpy\\_core\\__init__.py). Did you mean: 'asfarray'?\n",
      "]\n",
      "[autoreload of numpy.polynomial.hermite failed: Traceback (most recent call last):\n",
      "  File \"C:\\Users\\Harry\\AppData\\Roaming\\Python\\Python312\\site-packages\\IPython\\extensions\\autoreload.py\", line 276, in check\n",
      "    superreload(m, reload, self.old_objects)\n",
      "  File \"C:\\Users\\Harry\\AppData\\Roaming\\Python\\Python312\\site-packages\\IPython\\extensions\\autoreload.py\", line 475, in superreload\n",
      "    module = reload(module)\n",
      "             ^^^^^^^^^^^^^^\n",
      "  File \"D:\\anaconda\\Lib\\importlib\\__init__.py\", line 131, in reload\n",
      "    _bootstrap._exec(spec, module)\n",
      "  File \"<frozen importlib._bootstrap>\", line 866, in _exec\n",
      "  File \"<frozen importlib._bootstrap_external>\", line 995, in exec_module\n",
      "  File \"<frozen importlib._bootstrap>\", line 488, in _call_with_frames_removed\n",
      "  File \"D:\\anaconda\\Lib\\site-packages\\numpy\\polynomial\\hermite.py\", line 80, in <module>\n",
      "    from numpy.lib.array_utils import normalize_axis_index\n",
      "  File \"D:\\anaconda\\Lib\\site-packages\\numpy\\lib\\array_utils.py\", line 1, in <module>\n",
      "    from ._array_utils_impl import (\n",
      "  File \"D:\\anaconda\\Lib\\site-packages\\numpy\\lib\\_array_utils_impl.py\", line 4, in <module>\n",
      "    from numpy._core import asarray\n",
      "ImportError: cannot import name 'asarray' from 'numpy._core' (D:\\anaconda\\Lib\\site-packages\\numpy\\_core\\__init__.py). Did you mean: 'asfarray'?\n",
      "]\n",
      "[autoreload of numpy.polynomial.hermite_e failed: Traceback (most recent call last):\n",
      "  File \"C:\\Users\\Harry\\AppData\\Roaming\\Python\\Python312\\site-packages\\IPython\\extensions\\autoreload.py\", line 276, in check\n",
      "    superreload(m, reload, self.old_objects)\n",
      "  File \"C:\\Users\\Harry\\AppData\\Roaming\\Python\\Python312\\site-packages\\IPython\\extensions\\autoreload.py\", line 475, in superreload\n",
      "    module = reload(module)\n",
      "             ^^^^^^^^^^^^^^\n",
      "  File \"D:\\anaconda\\Lib\\importlib\\__init__.py\", line 131, in reload\n",
      "    _bootstrap._exec(spec, module)\n",
      "  File \"<frozen importlib._bootstrap>\", line 866, in _exec\n",
      "  File \"<frozen importlib._bootstrap_external>\", line 995, in exec_module\n",
      "  File \"<frozen importlib._bootstrap>\", line 488, in _call_with_frames_removed\n",
      "  File \"D:\\anaconda\\Lib\\site-packages\\numpy\\polynomial\\hermite_e.py\", line 80, in <module>\n",
      "    from numpy.lib.array_utils import normalize_axis_index\n",
      "  File \"D:\\anaconda\\Lib\\site-packages\\numpy\\lib\\array_utils.py\", line 1, in <module>\n",
      "    from ._array_utils_impl import (\n",
      "  File \"D:\\anaconda\\Lib\\site-packages\\numpy\\lib\\_array_utils_impl.py\", line 4, in <module>\n",
      "    from numpy._core import asarray\n",
      "ImportError: cannot import name 'asarray' from 'numpy._core' (D:\\anaconda\\Lib\\site-packages\\numpy\\_core\\__init__.py). Did you mean: 'asfarray'?\n",
      "]\n",
      "[autoreload of numpy.polynomial.laguerre failed: Traceback (most recent call last):\n",
      "  File \"C:\\Users\\Harry\\AppData\\Roaming\\Python\\Python312\\site-packages\\IPython\\extensions\\autoreload.py\", line 276, in check\n",
      "    superreload(m, reload, self.old_objects)\n",
      "  File \"C:\\Users\\Harry\\AppData\\Roaming\\Python\\Python312\\site-packages\\IPython\\extensions\\autoreload.py\", line 475, in superreload\n",
      "    module = reload(module)\n",
      "             ^^^^^^^^^^^^^^\n",
      "  File \"D:\\anaconda\\Lib\\importlib\\__init__.py\", line 131, in reload\n",
      "    _bootstrap._exec(spec, module)\n",
      "  File \"<frozen importlib._bootstrap>\", line 866, in _exec\n",
      "  File \"<frozen importlib._bootstrap_external>\", line 995, in exec_module\n",
      "  File \"<frozen importlib._bootstrap>\", line 488, in _call_with_frames_removed\n",
      "  File \"D:\\anaconda\\Lib\\site-packages\\numpy\\polynomial\\laguerre.py\", line 80, in <module>\n",
      "    from numpy.lib.array_utils import normalize_axis_index\n",
      "  File \"D:\\anaconda\\Lib\\site-packages\\numpy\\lib\\array_utils.py\", line 1, in <module>\n",
      "    from ._array_utils_impl import (\n",
      "  File \"D:\\anaconda\\Lib\\site-packages\\numpy\\lib\\_array_utils_impl.py\", line 4, in <module>\n",
      "    from numpy._core import asarray\n",
      "ImportError: cannot import name 'asarray' from 'numpy._core' (D:\\anaconda\\Lib\\site-packages\\numpy\\_core\\__init__.py). Did you mean: 'asfarray'?\n",
      "]\n",
      "[autoreload of numpy.ctypeslib failed: Traceback (most recent call last):\n",
      "  File \"C:\\Users\\Harry\\AppData\\Roaming\\Python\\Python312\\site-packages\\IPython\\extensions\\autoreload.py\", line 276, in check\n",
      "    superreload(m, reload, self.old_objects)\n",
      "  File \"C:\\Users\\Harry\\AppData\\Roaming\\Python\\Python312\\site-packages\\IPython\\extensions\\autoreload.py\", line 475, in superreload\n",
      "    module = reload(module)\n",
      "             ^^^^^^^^^^^^^^\n",
      "  File \"D:\\anaconda\\Lib\\importlib\\__init__.py\", line 131, in reload\n",
      "    _bootstrap._exec(spec, module)\n",
      "  File \"<frozen importlib._bootstrap>\", line 866, in _exec\n",
      "  File \"<frozen importlib._bootstrap_external>\", line 995, in exec_module\n",
      "  File \"<frozen importlib._bootstrap>\", line 488, in _call_with_frames_removed\n",
      "  File \"D:\\anaconda\\Lib\\site-packages\\numpy\\ctypeslib.py\", line 59, in <module>\n",
      "    from numpy._core.multiarray import _flagdict, flagsobj\n",
      "  File \"D:\\anaconda\\Lib\\site-packages\\numpy\\_core\\multiarray.py\", line 82, in <module>\n",
      "    @array_function_from_c_func_and_dispatcher(_multiarray_umath.empty_like)\n",
      "     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"D:\\anaconda\\Lib\\site-packages\\numpy\\_core\\overrides.py\", line 178, in decorator\n",
      "    return array_function_dispatch(\n",
      "           ^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"D:\\anaconda\\Lib\\site-packages\\numpy\\_core\\overrides.py\", line 158, in decorator\n",
      "    add_docstring(implementation, dispatcher.__doc__)\n",
      "RuntimeError: empty_like method already has a different docstring\n",
      "]\n",
      "[autoreload of numpy.ma.core failed: Traceback (most recent call last):\n",
      "  File \"C:\\Users\\Harry\\AppData\\Roaming\\Python\\Python312\\site-packages\\IPython\\extensions\\autoreload.py\", line 276, in check\n",
      "    superreload(m, reload, self.old_objects)\n",
      "  File \"C:\\Users\\Harry\\AppData\\Roaming\\Python\\Python312\\site-packages\\IPython\\extensions\\autoreload.py\", line 455, in superreload\n",
      "    if not append_obj(module, old_objects, name, obj):\n",
      "           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"C:\\Users\\Harry\\AppData\\Roaming\\Python\\Python312\\site-packages\\IPython\\extensions\\autoreload.py\", line 423, in append_obj\n",
      "    in_module = hasattr(obj, \"__module__\") and obj.__module__ == module.__name__\n",
      "                ^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"D:\\anaconda\\Lib\\site-packages\\numpy\\core\\umath.py\", line 2, in __getattr__\n",
      "    from numpy._core import umath\n",
      "  File \"D:\\anaconda\\Lib\\site-packages\\numpy\\_core\\umath.py\", line 15, in <module>\n",
      "    from ._multiarray_umath import (\n",
      "ImportError: cannot import name '_get_extobj_dict' from 'numpy.core._multiarray_umath' (D:\\anaconda\\Lib\\site-packages\\numpy\\core\\_multiarray_umath.cp312-win_amd64.pyd)\n",
      "]\n",
      "[autoreload of numpy.ma.extras failed: Traceback (most recent call last):\n",
      "  File \"C:\\Users\\Harry\\AppData\\Roaming\\Python\\Python312\\site-packages\\IPython\\extensions\\autoreload.py\", line 276, in check\n",
      "    superreload(m, reload, self.old_objects)\n",
      "  File \"C:\\Users\\Harry\\AppData\\Roaming\\Python\\Python312\\site-packages\\IPython\\extensions\\autoreload.py\", line 475, in superreload\n",
      "    module = reload(module)\n",
      "             ^^^^^^^^^^^^^^\n",
      "  File \"D:\\anaconda\\Lib\\importlib\\__init__.py\", line 131, in reload\n",
      "    _bootstrap._exec(spec, module)\n",
      "  File \"<frozen importlib._bootstrap>\", line 866, in _exec\n",
      "  File \"<frozen importlib._bootstrap_external>\", line 995, in exec_module\n",
      "  File \"<frozen importlib._bootstrap>\", line 488, in _call_with_frames_removed\n",
      "  File \"D:\\anaconda\\Lib\\site-packages\\numpy\\ma\\extras.py\", line 35, in <module>\n",
      "    from numpy.lib.array_utils import normalize_axis_index, normalize_axis_tuple\n",
      "  File \"D:\\anaconda\\Lib\\site-packages\\numpy\\lib\\array_utils.py\", line 1, in <module>\n",
      "    from ._array_utils_impl import (\n",
      "  File \"D:\\anaconda\\Lib\\site-packages\\numpy\\lib\\_array_utils_impl.py\", line 4, in <module>\n",
      "    from numpy._core import asarray\n",
      "ImportError: cannot import name 'asarray' from 'numpy._core' (D:\\anaconda\\Lib\\site-packages\\numpy\\_core\\__init__.py). Did you mean: 'asfarray'?\n",
      "]\n",
      "[autoreload of numpy._core failed: Traceback (most recent call last):\n",
      "  File \"C:\\Users\\Harry\\AppData\\Roaming\\Python\\Python312\\site-packages\\IPython\\extensions\\autoreload.py\", line 276, in check\n",
      "    superreload(m, reload, self.old_objects)\n",
      "  File \"C:\\Users\\Harry\\AppData\\Roaming\\Python\\Python312\\site-packages\\IPython\\extensions\\autoreload.py\", line 475, in superreload\n",
      "    module = reload(module)\n",
      "             ^^^^^^^^^^^^^^\n",
      "  File \"D:\\anaconda\\Lib\\importlib\\__init__.py\", line 131, in reload\n",
      "    _bootstrap._exec(spec, module)\n",
      "  File \"<frozen importlib._bootstrap>\", line 866, in _exec\n",
      "  File \"<frozen importlib._bootstrap_external>\", line 995, in exec_module\n",
      "  File \"<frozen importlib._bootstrap>\", line 488, in _call_with_frames_removed\n",
      "  File \"D:\\anaconda\\Lib\\site-packages\\numpy\\_core\\__init__.py\", line 23, in <module>\n",
      "    from . import multiarray\n",
      "  File \"D:\\anaconda\\Lib\\site-packages\\numpy\\_core\\multiarray.py\", line 82, in <module>\n",
      "    @array_function_from_c_func_and_dispatcher(_multiarray_umath.empty_like)\n",
      "     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"D:\\anaconda\\Lib\\site-packages\\numpy\\_core\\overrides.py\", line 178, in decorator\n",
      "    return array_function_dispatch(\n",
      "           ^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"D:\\anaconda\\Lib\\site-packages\\numpy\\_core\\overrides.py\", line 158, in decorator\n",
      "    add_docstring(implementation, dispatcher.__doc__)\n",
      "RuntimeError: empty_like method already has a different docstring\n",
      "]\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Collecting scikit-learn==0.24.2\n",
      "  Downloading scikit-learn-0.24.2.tar.gz (7.5 MB)\n",
      "     ---------------------------------------- 0.0/7.5 MB ? eta -:--:--\n",
      "     ---- ----------------------------------- 0.8/7.5 MB 16.9 MB/s eta 0:00:01\n",
      "     ------------- -------------------------- 2.6/7.5 MB 8.9 MB/s eta 0:00:01\n",
      "     -------------------- ------------------- 3.9/7.5 MB 7.8 MB/s eta 0:00:01\n",
      "     ----------------------- ---------------- 4.5/7.5 MB 5.8 MB/s eta 0:00:01\n",
      "     --------------------------- ------------ 5.2/7.5 MB 5.1 MB/s eta 0:00:01\n",
      "     --------------------------------- ------ 6.3/7.5 MB 5.0 MB/s eta 0:00:01\n",
      "     ---------------------------------------- 7.5/7.5 MB 5.7 MB/s eta 0:00:00\n",
      "  Installing build dependencies: started\n",
      "  Installing build dependencies: still running...\n",
      "  Installing build dependencies: finished with status 'done'\n",
      "  Getting requirements to build wheel: started\n",
      "  Getting requirements to build wheel: finished with status 'done'\n",
      "  Preparing metadata (pyproject.toml): started\n",
      "  Preparing metadata (pyproject.toml): finished with status 'error'\n",
      "Note: you may need to restart the kernel to use updated packages.\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "  error: subprocess-exited-with-error\n",
      "  \n",
      "  Preparing metadata (pyproject.toml) did not run successfully.\n",
      "  exit code: 1\n",
      "  \n",
      "  [20 lines of output]\n",
      "  <string>:17: DeprecationWarning: pkg_resources is deprecated as an API. See https://setuptools.pypa.io/en/latest/pkg_resources.html\n",
      "  Partial import of sklearn during the build process.\n",
      "  Traceback (most recent call last):\n",
      "    File \"D:\\anaconda\\Lib\\site-packages\\pip\\_vendor\\pyproject_hooks\\_in_process\\_in_process.py\", line 353, in <module>\n",
      "      main()\n",
      "    File \"D:\\anaconda\\Lib\\site-packages\\pip\\_vendor\\pyproject_hooks\\_in_process\\_in_process.py\", line 335, in main\n",
      "      json_out['return_val'] = hook(**hook_input['kwargs'])\n",
      "                               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "    File \"D:\\anaconda\\Lib\\site-packages\\pip\\_vendor\\pyproject_hooks\\_in_process\\_in_process.py\", line 149, in prepare_metadata_for_build_wheel\n",
      "      return hook(metadata_directory, config_settings)\n",
      "             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "    File \"C:\\Users\\Harry\\AppData\\Local\\Temp\\pip-build-env-w_gdxcy_\\overlay\\Lib\\site-packages\\setuptools\\build_meta.py\", line 373, in prepare_metadata_for_build_wheel\n",
      "      self.run_setup()\n",
      "    File \"C:\\Users\\Harry\\AppData\\Local\\Temp\\pip-build-env-w_gdxcy_\\overlay\\Lib\\site-packages\\setuptools\\build_meta.py\", line 503, in run_setup\n",
      "      super().run_setup(setup_script=setup_script)\n",
      "    File \"C:\\Users\\Harry\\AppData\\Local\\Temp\\pip-build-env-w_gdxcy_\\overlay\\Lib\\site-packages\\setuptools\\build_meta.py\", line 318, in run_setup\n",
      "      exec(code, locals())\n",
      "    File \"<string>\", line 301, in <module>\n",
      "    File \"<string>\", line 293, in setup_package\n",
      "  ModuleNotFoundError: No module named 'numpy.distutils'\n",
      "  [end of output]\n",
      "  \n",
      "  note: This error originates from a subprocess, and is likely not a problem with pip.\n",
      "error: metadata-generation-failed\n",
      "\n",
      "Encountered error while generating package metadata.\n",
      "\n",
      "See above for output.\n",
      "\n",
      "note: This is an issue with the package mentioned above, not pip.\n",
      "hint: See above for details.\n"
     ]
    }
   ],
   "source": [
    "pip install scikit-learn==0.24.2\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {
    "id": "-owEoJA9LsER"
   },
   "outputs": [],
   "source": [
    "import wandb\n",
    "import sklearn"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {
    "colab": {
     "base_uri": "https://localhost:8080/"
    },
    "id": "F50ju05EgX_B",
    "outputId": "acad4f92-0dce-4e92-b4af-d1423bba0dd3"
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "D:\\jupyter notebook\\JoJoGAN\\JoJoGAN\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "fatal: destination path 'JoJoGAN' already exists and is not an empty directory.\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Requirement already satisfied: tqdm in d:\\anaconda\\lib\\site-packages (4.66.4)\n",
      "Collecting gdown\n",
      "  Using cached gdown-5.2.0-py3-none-any.whl.metadata (5.8 kB)\n",
      "Collecting scikit-learn==0.22\n",
      "  Using cached scikit-learn-0.22.tar.gz (6.9 MB)\n",
      "  Preparing metadata (setup.py): started\n",
      "  Preparing metadata (setup.py): finished with status 'done'\n",
      "Requirement already satisfied: scipy in d:\\anaconda\\lib\\site-packages (1.13.1)\n",
      "Collecting lpips\n",
      "  Using cached lpips-0.1.4-py3-none-any.whl.metadata (10 kB)\n",
      "Requirement already satisfied: dlib in d:\\anaconda\\lib\\site-packages (19.24.6)\n",
      "Requirement already satisfied: opencv-python in d:\\anaconda\\lib\\site-packages (4.10.0.84)\n",
      "Requirement already satisfied: wandb in d:\\anaconda\\lib\\site-packages (0.17.7)\n",
      "Requirement already satisfied: numpy>=1.11.0 in d:\\anaconda\\lib\\site-packages (from scikit-learn==0.22) (1.26.4)\n",
      "Requirement already satisfied: joblib>=0.11 in d:\\anaconda\\lib\\site-packages (from scikit-learn==0.22) (1.4.2)\n",
      "Requirement already satisfied: colorama in c:\\users\\harry\\appdata\\roaming\\python\\python312\\site-packages (from tqdm) (0.4.6)\n",
      "Requirement already satisfied: beautifulsoup4 in d:\\anaconda\\lib\\site-packages (from gdown) (4.12.3)\n",
      "Requirement already satisfied: filelock in d:\\anaconda\\lib\\site-packages (from gdown) (3.13.1)\n",
      "Requirement already satisfied: requests[socks] in d:\\anaconda\\lib\\site-packages (from gdown) (2.32.2)\n",
      "Requirement already satisfied: torch>=0.4.0 in d:\\anaconda\\lib\\site-packages (from lpips) (2.4.0+cu118)\n",
      "Requirement already satisfied: torchvision>=0.2.1 in d:\\anaconda\\lib\\site-packages (from lpips) (0.19.0+cu118)\n",
      "Requirement already satisfied: click!=8.0.0,>=7.1 in d:\\anaconda\\lib\\site-packages (from wandb) (8.1.7)\n",
      "Requirement already satisfied: docker-pycreds>=0.4.0 in d:\\anaconda\\lib\\site-packages (from wandb) (0.4.0)\n",
      "Requirement already satisfied: gitpython!=3.1.29,>=1.0.0 in d:\\anaconda\\lib\\site-packages (from wandb) (3.1.37)\n",
      "Requirement already satisfied: platformdirs in c:\\users\\harry\\appdata\\roaming\\python\\python312\\site-packages (from wandb) (4.2.2)\n",
      "Requirement already satisfied: protobuf!=4.21.0,<6,>=3.19.0 in d:\\anaconda\\lib\\site-packages (from wandb) (3.20.3)\n",
      "Requirement already satisfied: psutil>=5.0.0 in c:\\users\\harry\\appdata\\roaming\\python\\python312\\site-packages (from wandb) (6.0.0)\n",
      "Requirement already satisfied: pyyaml in d:\\anaconda\\lib\\site-packages (from wandb) (6.0.1)\n",
      "Requirement already satisfied: sentry-sdk>=1.0.0 in d:\\anaconda\\lib\\site-packages (from wandb) (2.13.0)\n",
      "Requirement already satisfied: setproctitle in d:\\anaconda\\lib\\site-packages (from wandb) (1.3.3)\n",
      "Requirement already satisfied: setuptools in d:\\anaconda\\lib\\site-packages (from wandb) (74.0.0)\n",
      "Requirement already satisfied: six>=1.4.0 in c:\\users\\harry\\appdata\\roaming\\python\\python312\\site-packages (from docker-pycreds>=0.4.0->wandb) (1.16.0)\n",
      "Requirement already satisfied: gitdb<5,>=4.0.1 in d:\\anaconda\\lib\\site-packages (from gitpython!=3.1.29,>=1.0.0->wandb) (4.0.7)\n",
      "Requirement already satisfied: charset-normalizer<4,>=2 in d:\\anaconda\\lib\\site-packages (from requests[socks]->gdown) (2.0.4)\n",
      "Requirement already satisfied: idna<4,>=2.5 in d:\\anaconda\\lib\\site-packages (from requests[socks]->gdown) (3.7)\n",
      "Requirement already satisfied: urllib3<3,>=1.21.1 in d:\\anaconda\\lib\\site-packages (from requests[socks]->gdown) (2.2.2)\n",
      "Requirement already satisfied: certifi>=2017.4.17 in d:\\anaconda\\lib\\site-packages (from requests[socks]->gdown) (2024.6.2)\n",
      "Requirement already satisfied: typing-extensions>=4.8.0 in d:\\anaconda\\lib\\site-packages (from torch>=0.4.0->lpips) (4.11.0)\n",
      "Requirement already satisfied: sympy in d:\\anaconda\\lib\\site-packages (from torch>=0.4.0->lpips) (1.12)\n",
      "Requirement already satisfied: networkx in d:\\anaconda\\lib\\site-packages (from torch>=0.4.0->lpips) (3.2.1)\n",
      "Requirement already satisfied: jinja2 in d:\\anaconda\\lib\\site-packages (from torch>=0.4.0->lpips) (3.1.4)\n",
      "Requirement already satisfied: fsspec in d:\\anaconda\\lib\\site-packages (from torch>=0.4.0->lpips) (2024.3.1)\n",
      "Requirement already satisfied: pillow!=8.3.*,>=5.3.0 in d:\\anaconda\\lib\\site-packages (from torchvision>=0.2.1->lpips) (10.3.0)\n",
      "Requirement already satisfied: soupsieve>1.2 in d:\\anaconda\\lib\\site-packages (from beautifulsoup4->gdown) (2.5)\n",
      "Requirement already satisfied: PySocks!=1.5.7,>=1.5.6 in d:\\anaconda\\lib\\site-packages (from requests[socks]->gdown) (1.7.1)\n",
      "Requirement already satisfied: smmap<5,>=3.0.1 in d:\\anaconda\\lib\\site-packages (from gitdb<5,>=4.0.1->gitpython!=3.1.29,>=1.0.0->wandb) (4.0.0)\n",
      "Requirement already satisfied: MarkupSafe>=2.0 in d:\\anaconda\\lib\\site-packages (from jinja2->torch>=0.4.0->lpips) (2.1.3)\n",
      "Requirement already satisfied: mpmath>=0.19 in d:\\anaconda\\lib\\site-packages (from sympy->torch>=0.4.0->lpips) (1.3.0)\n",
      "Using cached gdown-5.2.0-py3-none-any.whl (18 kB)\n",
      "Using cached lpips-0.1.4-py3-none-any.whl (53 kB)\n",
      "Building wheels for collected packages: scikit-learn\n",
      "  Building wheel for scikit-learn (setup.py): started\n",
      "  Building wheel for scikit-learn (setup.py): finished with status 'error'\n",
      "  Running setup.py clean for scikit-learn\n",
      "Failed to build scikit-learn\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "  error: subprocess-exited-with-error\n",
      "  \n",
      "  python setup.py bdist_wheel did not run successfully.\n",
      "  exit code: 1\n",
      "  \n",
      "  [11 lines of output]\n",
      "  C:\\Users\\Harry\\AppData\\Local\\Temp\\pip-install-i2ayqo8b\\scikit-learn_fcbeb724cceb496692b5d4beff649177\\setup.py:12: DeprecationWarning: pkg_resources is deprecated as an API. See https://setuptools.pypa.io/en/latest/pkg_resources.html\n",
      "    from pkg_resources import parse_version\n",
      "  Partial import of sklearn during the build process.\n",
      "  Traceback (most recent call last):\n",
      "    File \"<string>\", line 2, in <module>\n",
      "    File \"<pip-setuptools-caller>\", line 34, in <module>\n",
      "    File \"C:\\Users\\Harry\\AppData\\Local\\Temp\\pip-install-i2ayqo8b\\scikit-learn_fcbeb724cceb496692b5d4beff649177\\setup.py\", line 303, in <module>\n",
      "      setup_package()\n",
      "    File \"C:\\Users\\Harry\\AppData\\Local\\Temp\\pip-install-i2ayqo8b\\scikit-learn_fcbeb724cceb496692b5d4beff649177\\setup.py\", line 295, in setup_package\n",
      "      from numpy.distutils.core import setup\n",
      "  ModuleNotFoundError: No module named 'numpy.distutils'\n",
      "  [end of output]\n",
      "  \n",
      "  note: This error originates from a subprocess, and is likely not a problem with pip.\n",
      "  ERROR: Failed building wheel for scikit-learn\n",
      "ERROR: ERROR: Failed to build installable wheels for some pyproject.toml based projects (scikit-learn)\n",
      "  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current\n",
      "                                 Dload  Upload   Total   Spent    Left  Speed\n",
      "\n",
      "  0     0    0     0    0     0      0      0 --:--:-- --:--:-- --:--:--     0\n",
      "  0     0    0     0    0     0      0      0 --:--:-- --:--:-- --:--:--     0\n",
      "\n",
      "  0     0    0     0    0     0      0      0 --:--:--  0:00:01 --:--:--     0\n",
      " 13  241k   13 32768    0     0  15031      0  0:00:16  0:00:02  0:00:14 72495\n",
      "100  241k  100  241k    0     0   108k      0  0:00:02  0:00:02 --:--:--  481k\n",
      "tar: could not chdir to 'C:\\ninja'\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "The autoreload extension is already loaded. To reload it, use:\n",
      "  %reload_ext autoreload\n"
     ]
    },
    {
     "ename": "ImportError",
     "evalue": "DLL load failed while importing fused: The specified module could not be found.",
     "output_type": "error",
     "traceback": [
      "\u001b[1;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[1;31mImportError\u001b[0m                               Traceback (most recent call last)",
      "Cell \u001b[1;32mIn[4], line 28\u001b[0m\n\u001b[0;32m     26\u001b[0m \u001b[38;5;28;01mfrom\u001b[39;00m \u001b[38;5;21;01mtqdm\u001b[39;00m \u001b[38;5;28;01mimport\u001b[39;00m tqdm\n\u001b[0;32m     27\u001b[0m \u001b[38;5;28;01mimport\u001b[39;00m \u001b[38;5;21;01mwandb\u001b[39;00m\n\u001b[1;32m---> 28\u001b[0m \u001b[38;5;28;01mfrom\u001b[39;00m \u001b[38;5;21;01mmodel\u001b[39;00m \u001b[38;5;28;01mimport\u001b[39;00m \u001b[38;5;241m*\u001b[39m\n\u001b[0;32m     29\u001b[0m \u001b[38;5;28;01mfrom\u001b[39;00m \u001b[38;5;21;01me4e_projection\u001b[39;00m \u001b[38;5;28;01mimport\u001b[39;00m projection \u001b[38;5;28;01mas\u001b[39;00m e4e_projection\n\u001b[0;32m     31\u001b[0m \u001b[38;5;66;03m# Handling Google Colab-specific code\u001b[39;00m\n\u001b[0;32m     32\u001b[0m \u001b[38;5;66;03m# You can remove this block if you are not using Colab\u001b[39;00m\n",
      "File \u001b[1;32mD:\\jupyter notebook\\JoJoGAN\\JoJoGAN\\model.py:13\u001b[0m\n\u001b[0;32m     11\u001b[0m \u001b[38;5;28;01mfrom\u001b[39;00m \u001b[38;5;21;01mop\u001b[39;00m \u001b[38;5;28;01mimport\u001b[39;00m conv2d_gradfix\n\u001b[0;32m     12\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m torch\u001b[38;5;241m.\u001b[39mcuda\u001b[38;5;241m.\u001b[39mis_available():\n\u001b[1;32m---> 13\u001b[0m     \u001b[38;5;28;01mfrom\u001b[39;00m \u001b[38;5;21;01mop\u001b[39;00m\u001b[38;5;21;01m.\u001b[39;00m\u001b[38;5;21;01mfused_act\u001b[39;00m \u001b[38;5;28;01mimport\u001b[39;00m FusedLeakyReLU, fused_leaky_relu\n\u001b[0;32m     14\u001b[0m     \u001b[38;5;28;01mfrom\u001b[39;00m \u001b[38;5;21;01mop\u001b[39;00m\u001b[38;5;21;01m.\u001b[39;00m\u001b[38;5;21;01mupfirdn2d\u001b[39;00m \u001b[38;5;28;01mimport\u001b[39;00m upfirdn2d\n\u001b[0;32m     15\u001b[0m \u001b[38;5;28;01melse\u001b[39;00m:\n",
      "File \u001b[1;32mD:\\jupyter notebook\\JoJoGAN\\op\\fused_act.py:11\u001b[0m\n\u001b[0;32m      7\u001b[0m \u001b[38;5;28;01mfrom\u001b[39;00m \u001b[38;5;21;01mtorch\u001b[39;00m\u001b[38;5;21;01m.\u001b[39;00m\u001b[38;5;21;01mutils\u001b[39;00m\u001b[38;5;21;01m.\u001b[39;00m\u001b[38;5;21;01mcpp_extension\u001b[39;00m \u001b[38;5;28;01mimport\u001b[39;00m load\n\u001b[0;32m     10\u001b[0m module_path \u001b[38;5;241m=\u001b[39m os\u001b[38;5;241m.\u001b[39mpath\u001b[38;5;241m.\u001b[39mdirname(\u001b[38;5;18m__file__\u001b[39m)\n\u001b[1;32m---> 11\u001b[0m fused \u001b[38;5;241m=\u001b[39m \u001b[43mload\u001b[49m\u001b[43m(\u001b[49m\n\u001b[0;32m     12\u001b[0m \u001b[43m    \u001b[49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[38;5;124;43mfused\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[43m,\u001b[49m\n\u001b[0;32m     13\u001b[0m \u001b[43m    \u001b[49m\u001b[43msources\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43m[\u001b[49m\n\u001b[0;32m     14\u001b[0m \u001b[43m        \u001b[49m\u001b[43mos\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mpath\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mjoin\u001b[49m\u001b[43m(\u001b[49m\u001b[43mmodule_path\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[38;5;124;43mfused_bias_act.cpp\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[43m)\u001b[49m\u001b[43m,\u001b[49m\n\u001b[0;32m     15\u001b[0m \u001b[43m        \u001b[49m\u001b[43mos\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mpath\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mjoin\u001b[49m\u001b[43m(\u001b[49m\u001b[43mmodule_path\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[38;5;124;43mfused_bias_act_kernel.cu\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[43m)\u001b[49m\u001b[43m,\u001b[49m\n\u001b[0;32m     16\u001b[0m \u001b[43m    \u001b[49m\u001b[43m]\u001b[49m\u001b[43m,\u001b[49m\n\u001b[0;32m     17\u001b[0m \u001b[43m)\u001b[49m\n\u001b[0;32m     20\u001b[0m \u001b[38;5;28;01mclass\u001b[39;00m \u001b[38;5;21;01mFusedLeakyReLUFunctionBackward\u001b[39;00m(Function):\n\u001b[0;32m     21\u001b[0m     \u001b[38;5;129m@staticmethod\u001b[39m\n\u001b[0;32m     22\u001b[0m     \u001b[38;5;28;01mdef\u001b[39;00m \u001b[38;5;21mforward\u001b[39m(ctx, grad_output, out, bias, negative_slope, scale):\n",
      "File \u001b[1;32mD:\\anaconda\\Lib\\site-packages\\torch\\utils\\cpp_extension.py:1312\u001b[0m, in \u001b[0;36mload\u001b[1;34m(name, sources, extra_cflags, extra_cuda_cflags, extra_ldflags, extra_include_paths, build_directory, verbose, with_cuda, is_python_module, is_standalone, keep_intermediates)\u001b[0m\n\u001b[0;32m   1220\u001b[0m \u001b[38;5;28;01mdef\u001b[39;00m \u001b[38;5;21mload\u001b[39m(name,\n\u001b[0;32m   1221\u001b[0m          sources: Union[\u001b[38;5;28mstr\u001b[39m, List[\u001b[38;5;28mstr\u001b[39m]],\n\u001b[0;32m   1222\u001b[0m          extra_cflags\u001b[38;5;241m=\u001b[39m\u001b[38;5;28;01mNone\u001b[39;00m,\n\u001b[1;32m   (...)\u001b[0m\n\u001b[0;32m   1230\u001b[0m          is_standalone\u001b[38;5;241m=\u001b[39m\u001b[38;5;28;01mFalse\u001b[39;00m,\n\u001b[0;32m   1231\u001b[0m          keep_intermediates\u001b[38;5;241m=\u001b[39m\u001b[38;5;28;01mTrue\u001b[39;00m):\n\u001b[0;32m   1232\u001b[0m \u001b[38;5;250m    \u001b[39m\u001b[38;5;124;03m\"\"\"\u001b[39;00m\n\u001b[0;32m   1233\u001b[0m \u001b[38;5;124;03m    Load a PyTorch C++ extension just-in-time (JIT).\u001b[39;00m\n\u001b[0;32m   1234\u001b[0m \n\u001b[1;32m   (...)\u001b[0m\n\u001b[0;32m   1310\u001b[0m \u001b[38;5;124;03m        ...     verbose=True)\u001b[39;00m\n\u001b[0;32m   1311\u001b[0m \u001b[38;5;124;03m    \"\"\"\u001b[39;00m\n\u001b[1;32m-> 1312\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[43m_jit_compile\u001b[49m\u001b[43m(\u001b[49m\n\u001b[0;32m   1313\u001b[0m \u001b[43m        \u001b[49m\u001b[43mname\u001b[49m\u001b[43m,\u001b[49m\n\u001b[0;32m   1314\u001b[0m \u001b[43m        \u001b[49m\u001b[43m[\u001b[49m\u001b[43msources\u001b[49m\u001b[43m]\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;28;43;01mif\u001b[39;49;00m\u001b[43m \u001b[49m\u001b[38;5;28;43misinstance\u001b[39;49m\u001b[43m(\u001b[49m\u001b[43msources\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;28;43mstr\u001b[39;49m\u001b[43m)\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;28;43;01melse\u001b[39;49;00m\u001b[43m \u001b[49m\u001b[43msources\u001b[49m\u001b[43m,\u001b[49m\n\u001b[0;32m   1315\u001b[0m \u001b[43m        \u001b[49m\u001b[43mextra_cflags\u001b[49m\u001b[43m,\u001b[49m\n\u001b[0;32m   1316\u001b[0m \u001b[43m        \u001b[49m\u001b[43mextra_cuda_cflags\u001b[49m\u001b[43m,\u001b[49m\n\u001b[0;32m   1317\u001b[0m \u001b[43m        \u001b[49m\u001b[43mextra_ldflags\u001b[49m\u001b[43m,\u001b[49m\n\u001b[0;32m   1318\u001b[0m \u001b[43m        \u001b[49m\u001b[43mextra_include_paths\u001b[49m\u001b[43m,\u001b[49m\n\u001b[0;32m   1319\u001b[0m \u001b[43m        \u001b[49m\u001b[43mbuild_directory\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;129;43;01mor\u001b[39;49;00m\u001b[43m \u001b[49m\u001b[43m_get_build_directory\u001b[49m\u001b[43m(\u001b[49m\u001b[43mname\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mverbose\u001b[49m\u001b[43m)\u001b[49m\u001b[43m,\u001b[49m\n\u001b[0;32m   1320\u001b[0m \u001b[43m        \u001b[49m\u001b[43mverbose\u001b[49m\u001b[43m,\u001b[49m\n\u001b[0;32m   1321\u001b[0m \u001b[43m        \u001b[49m\u001b[43mwith_cuda\u001b[49m\u001b[43m,\u001b[49m\n\u001b[0;32m   1322\u001b[0m \u001b[43m        \u001b[49m\u001b[43mis_python_module\u001b[49m\u001b[43m,\u001b[49m\n\u001b[0;32m   1323\u001b[0m \u001b[43m        \u001b[49m\u001b[43mis_standalone\u001b[49m\u001b[43m,\u001b[49m\n\u001b[0;32m   1324\u001b[0m \u001b[43m        \u001b[49m\u001b[43mkeep_intermediates\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mkeep_intermediates\u001b[49m\u001b[43m)\u001b[49m\n",
      "File \u001b[1;32mD:\\anaconda\\Lib\\site-packages\\torch\\utils\\cpp_extension.py:1747\u001b[0m, in \u001b[0;36m_jit_compile\u001b[1;34m(name, sources, extra_cflags, extra_cuda_cflags, extra_ldflags, extra_include_paths, build_directory, verbose, with_cuda, is_python_module, is_standalone, keep_intermediates)\u001b[0m\n\u001b[0;32m   1744\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m is_standalone:\n\u001b[0;32m   1745\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m _get_exec_path(name, build_directory)\n\u001b[1;32m-> 1747\u001b[0m \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[43m_import_module_from_library\u001b[49m\u001b[43m(\u001b[49m\u001b[43mname\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mbuild_directory\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mis_python_module\u001b[49m\u001b[43m)\u001b[49m\n",
      "File \u001b[1;32mD:\\anaconda\\Lib\\site-packages\\torch\\utils\\cpp_extension.py:2141\u001b[0m, in \u001b[0;36m_import_module_from_library\u001b[1;34m(module_name, path, is_python_module)\u001b[0m\n\u001b[0;32m   2139\u001b[0m spec \u001b[38;5;241m=\u001b[39m importlib\u001b[38;5;241m.\u001b[39mutil\u001b[38;5;241m.\u001b[39mspec_from_file_location(module_name, filepath)\n\u001b[0;32m   2140\u001b[0m \u001b[38;5;28;01massert\u001b[39;00m spec \u001b[38;5;129;01mis\u001b[39;00m \u001b[38;5;129;01mnot\u001b[39;00m \u001b[38;5;28;01mNone\u001b[39;00m\n\u001b[1;32m-> 2141\u001b[0m module \u001b[38;5;241m=\u001b[39m \u001b[43mimportlib\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mutil\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mmodule_from_spec\u001b[49m\u001b[43m(\u001b[49m\u001b[43mspec\u001b[49m\u001b[43m)\u001b[49m\n\u001b[0;32m   2142\u001b[0m \u001b[38;5;28;01massert\u001b[39;00m \u001b[38;5;28misinstance\u001b[39m(spec\u001b[38;5;241m.\u001b[39mloader, importlib\u001b[38;5;241m.\u001b[39mabc\u001b[38;5;241m.\u001b[39mLoader)\n\u001b[0;32m   2143\u001b[0m spec\u001b[38;5;241m.\u001b[39mloader\u001b[38;5;241m.\u001b[39mexec_module(module)\n",
      "\u001b[1;31mImportError\u001b[0m: DLL load failed while importing fused: The specified module could not be found."
     ]
    }
   ],
   "source": [
    "# Setup. This will take a few minutes.\n",
    "!git clone https://github.com/mchong6/JoJoGAN.git\n",
    "%cd JoJoGAN\n",
    "!pip install tqdm gdown scikit-learn==0.22 scipy lpips dlib opencv-python wandb\n",
    "\n",
    "# Download the Ninja build tool for Windows\n",
    "!curl -L -o ninja-win.zip https://github.com/ninja-build/ninja/releases/download/v1.8.2/ninja-win.zip\n",
    "!tar -xf ninja-win.zip -C C:\\ninja\n",
    "!set PATH=%PATH%;C:\\ninja\n",
    "\n",
    "%load_ext autoreload\n",
    "%autoreload 2\n",
    "\n",
    "import torch\n",
    "torch.backends.cudnn.benchmark = True\n",
    "from torchvision import transforms, utils\n",
    "from util import *\n",
    "from PIL import Image\n",
    "import math\n",
    "import random\n",
    "import os\n",
    "\n",
    "import numpy as np\n",
    "from torch import nn, autograd, optim\n",
    "from torch.nn import functional as F\n",
    "from tqdm import tqdm\n",
    "import wandb\n",
    "from model import *\n",
    "from e4e_projection import projection as e4e_projection\n",
    "\n",
    "# Handling Google Colab-specific code\n",
    "# You can remove this block if you are not using Colab\n",
    "try:\n",
    "    from google.colab import files\n",
    "    from pydrive.auth import GoogleAuth\n",
    "    from pydrive.drive import GoogleDrive\n",
    "    from google.colab import auth\n",
    "    from oauth2client.client import GoogleCredentials\n",
    "except ImportError:\n",
    "    pass  # Google Colab modules not available on non-Colab environments\n",
    "\n",
    "# Create directories if they do not exist\n",
    "os.makedirs('inversion_codes', exist_ok=True)\n",
    "os.makedirs('style_images', exist_ok=True)\n",
    "os.makedirs('style_images_aligned', exist_ok=True)\n",
    "os.makedirs('models', exist_ok=True)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "id": "D7_hxx0jyon2"
   },
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "colab": {
     "base_uri": "https://localhost:8080/"
    },
    "id": "AgfLsIELgsLq",
    "outputId": "8f4aaea8-2cab-4ab1-87e2-3d4945a6b0b7"
   },
   "outputs": [],
   "source": [
    "#@title Download models\n",
    "#@markdown You may optionally enable downloads with pydrive in order to authenticate and avoid drive download limits.\n",
    "download_with_pydrive = True #@param {type:\"boolean\"}\n",
    "device = 'cuda' #@param ['cuda', 'cpu']\n",
    "\n",
    "!wget http:\\\\dlib.net\\files\\shape_predictor_68_face_landmarks.dat.bz2\n",
    "!bzip2 -dk shape_predictor_68_face_landmarks.dat.bz2\n",
    "!mv shape_predictor_68_face_landmarks.dat models\\dlibshape_predictor_68_face_landmarks.dat\n",
    "%matplotlib inline\n",
    "\n",
    "\n",
    "drive_ids = {\n",
    "    \"stylegan2-ffhq-config-f.pt\": \"1Yr7KuD959btpmcKGAUsbAk5rPjX2MytK\",\n",
    "    \"e4e_ffhq_encode.pt\": \"1o6ijA3PkcewZvwJJ73dJ0fxhndn0nnh7\",\n",
    "    \"restyle_psp_ffhq_encode.pt\": \"1nbxCIVw9H3YnQsoIPykNEFwWJnHVHlVd\",\n",
    "    \"arcane_caitlyn.pt\": \"1gOsDTiTPcENiFOrhmkkxJcTURykW1dRc\",\n",
    "    \"arcane_caitlyn_preserve_color.pt\": \"1cUTyjU-q98P75a8THCaO545RTwpVV-aH\",\n",
    "    \"arcane_jinx_preserve_color.pt\": \"1jElwHxaYPod5Itdy18izJk49K1nl4ney\",\n",
    "    \"arcane_jinx.pt\": \"1quQ8vPjYpUiXM4k1_KIwP4EccOefPpG_\",\n",
    "    \"arcane_multi_preserve_color.pt\": \"1enJgrC08NpWpx2XGBmLt1laimjpGCyfl\",\n",
    "    \"arcane_multi.pt\": \"15V9s09sgaw-zhKp116VHigf5FowAy43f\",\n",
    "    \"sketch_multi.pt\": \"1GdaeHGBGjBAFsWipTL0y-ssUiAqk8AxD\",\n",
    "    \"disney.pt\": \"1zbE2upakFUAx8ximYnLofFwfT8MilqJA\",\n",
    "    \"disney_preserve_color.pt\": \"1Bnh02DjfvN_Wm8c4JdOiNV4q9J7Z_tsi\",\n",
    "    \"jojo.pt\": \"13cR2xjIBj8Ga5jMO7gtxzIJj2PDsBYK4\",\n",
    "    \"jojo_preserve_color.pt\": \"1ZRwYLRytCEKi__eT2Zxv1IlV6BGVQ_K2\",\n",
    "    \"jojo_yasuho.pt\": \"1grZT3Gz1DLzFoJchAmoj3LoM9ew9ROX_\",\n",
    "    \"jojo_yasuho_preserve_color.pt\": \"1SKBu1h0iRNyeKBnya_3BBmLr4pkPeg_L\",\n",
    "    \"art.pt\": \"1a0QDEHwXQ6hE_FcYEyNMuv5r5UnRQLKT\",\n",
    "}\n",
    "\n",
    "# from StyelGAN-NADA\n",
    "class Downloader(object):\n",
    "    def __init__(self, use_pydrive):\n",
    "        self.use_pydrive = use_pydrive\n",
    "\n",
    "        if self.use_pydrive:\n",
    "            self.authenticate()\n",
    "\n",
    "    def authenticate(self):\n",
    "        auth.authenticate_user()\n",
    "        gauth = GoogleAuth()\n",
    "        gauth.credentials = GoogleCredentials.get_application_default()\n",
    "        self.drive = GoogleDrive(gauth)\n",
    "\n",
    "    def download_file(self, file_name):\n",
    "        file_dst = os.path.join('models', file_name)\n",
    "        file_id = drive_ids[file_name]\n",
    "        if not os.path.exists(file_dst):\n",
    "            print(f'Downloading {file_name}')\n",
    "            if self.use_pydrive:\n",
    "                downloaded = self.drive.CreateFile({'id':file_id})\n",
    "                downloaded.FetchMetadata(fetch_all=True)\n",
    "                downloaded.GetContentFile(file_dst)\n",
    "            else:\n",
    "                !gdown --id $file_id -O $file_dst\n",
    "\n",
    "\n",
    "\n",
    "downloader = Downloader(download_with_pydrive)\n",
    "\n",
    "downloader.download_file('stylegan2-ffhq-config-f.pt')\n",
    "downloader.download_file('e4e_ffhq_encode.pt')\n",
    "\n",
    "latent_dim = 512\n",
    "\n",
    "# Load original generator\n",
    "original_generator = Generator(1024, latent_dim, 8, 2).to(device)\n",
    "ckpt = torch.load('models\\stylegan2-ffhq-config-f.pt', map_location=lambda storage, loc: storage)\n",
    "original_generator.load_state_dict(ckpt[\"g_ema\"], strict=False)\n",
    "mean_latent = original_generator.mean_latent(10000)\n",
    "\n",
    "# to be finetuned generator\n",
    "generator = deepcopy(original_generator)\n",
    "\n",
    "\n",
    "transform = transforms.Compose(\n",
    "    [\n",
    "        transforms.Resize((1024, 1024)),\n",
    "        transforms.ToTensor(),\n",
    "        transforms.Normalize((0.5, 0.5, 0.5), (0.5, 0.5, 0.5)),\n",
    "    ]\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "colab": {
     "base_uri": "https://localhost:8080/",
     "height": 743
    },
    "id": "vCZPSATBFzMI",
    "outputId": "f9c77035-21fe-4558-94e4-65fd8793852c"
   },
   "outputs": [],
   "source": [
    "plt.rcParams['figure.dpi'] = 150\n",
    "\n",
    "#@title Choose input face\n",
    "#@markdown Add your own image to the test_input directory and put the name here\n",
    "filename = 'chris_hemsworth.jpeg' #@param {type:\"string\"}\n",
    "filepath = f'\\content\\JoJoGAN\\test_input\\chris_hemsworth.jpeg'\n",
    "\n",
    "# uploaded = files.upload()\n",
    "# filepath = list(uploaded.keys())[0]\n",
    "name = strip_path_extension(filepath)+'.pt'\n",
    "\n",
    "# aligns and crops face\n",
    "aligned_face = align_face(filepath)\n",
    "\n",
    "# my_w = restyle_projection(aligned_face, name, device, n_iters=1).unsqueeze(0)\n",
    "my_w = e4e_projection(aligned_face, name, device).unsqueeze(0)\n",
    "\n",
    "display_image(aligned_face, title='Aligned face')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "id": "gPMywfSGJUsI"
   },
   "source": [
    "# Use pretrained styles"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "colab": {
     "base_uri": "https://localhost:8080/",
     "height": 796
    },
    "id": "LCLWiXoXwcJb",
    "outputId": "43b7b7f4-3abc-4414-8c0e-a10e1e12e25e"
   },
   "outputs": [],
   "source": [
    "plt.rcParams['figure.dpi'] = 150\n",
    "pretrained = 'art' #@param ['art', 'arcane_multi', 'sketch_multi', 'arcane_jinx', 'arcane_caitlyn', 'jojo_yasuho', 'jojo', 'disney']\n",
    "#@markdown Preserve color tries to preserve color of original image by limiting family of allowable transformations. Otherwise, the stylized image will inherit the colors of the reference images, leading to heavier stylizations.\n",
    "preserve_color = False #@param{type:\"boolean\"}\n",
    "\n",
    "if preserve_color:\n",
    "    ckpt = f'{pretrained}_preserve_color.pt'\n",
    "else:\n",
    "    ckpt = f'{pretrained}.pt'\n",
    "\n",
    "# load base version if preserve_color version not available\n",
    "try:\n",
    "    downloader.download_file(ckpt)\n",
    "except:\n",
    "    ckpt = f'{pretrained}.pt'\n",
    "    downloader.download_file(ckpt)\n",
    "\n",
    "ckpt = torch.load(os.path.join('models', ckpt), map_location=lambda storage, loc: storage)\n",
    "generator.load_state_dict(ckpt[\"g\"], strict=False)\n",
    "\n",
    "#@title Generate results\n",
    "n_sample =  5#@param {type:\"number\"}\n",
    "seed = 3000 #@param {type:\"number\"}\n",
    "\n",
    "torch.manual_seed(seed)\n",
    "with torch.no_grad():\n",
    "    generator.eval()\n",
    "    z = torch.randn(n_sample, latent_dim, device=device)\n",
    "\n",
    "    original_sample = original_generator([z], truncation=0.7, truncation_latent=mean_latent)\n",
    "    sample = generator([z], truncation=0.7, truncation_latent=mean_latent)\n",
    "\n",
    "    original_my_sample = original_generator(my_w, input_is_latent=True)\n",
    "    my_sample = generator(my_w, input_is_latent=True)\n",
    "\n",
    "# display reference images\n",
    "if pretrained == 'arcane_multi':\n",
    "    style_path = f'style_images_aligned\\arcane_jinx.png'\n",
    "elif pretrained == 'sketch_multi':\n",
    "    style_path = f'style_images_aligned\\sketch.png'\n",
    "else:\n",
    "    style_path = f'style_images_aligned\\{pretrained}.png'\n",
    "\n",
    "style_image = transform(Image.open(style_path)).unsqueeze(0).to(device)\n",
    "face = transform(aligned_face).unsqueeze(0).to(device)\n",
    "\n",
    "my_output = torch.cat([style_image, face, my_sample], 0)\n",
    "display_image(utils.make_grid(my_output, normalize=True), title='My sample')\n",
    "\n",
    "output = torch.cat([original_sample, sample], 0)\n",
    "display_image(utils.make_grid(output, normalize=True, nrow=n_sample), title='Random samples')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "id": "VZdYrfVcJdqa"
   },
   "source": [
    "# Train with your own style images"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "colab": {
     "base_uri": "https://localhost:8080/",
     "height": 337
    },
    "id": "hgfk78K1ikb8",
    "outputId": "3514d297-ca71-47c1-ed60-7190e1064aaf"
   },
   "outputs": [],
   "source": [
    "#@markdown Upload your own style images into the style_images folder and type it into the field in the following format without the directory name. Upload multiple style images to do multi-shot image translation\n",
    "names = ['sketch.jpeg', 'sketch2.jpeg', 'sketch3.jpeg', 'sketch4.jpeg'] #@param {type:\"raw\"}\n",
    "\n",
    "targets = []\n",
    "latents = []\n",
    "\n",
    "for name in names:\n",
    "    style_path = os.path.join('style_images', name)\n",
    "    assert os.path.exists(style_path), f\"{style_path} does not exist!\"\n",
    "\n",
    "    name = strip_path_extension(name)\n",
    "\n",
    "    # crop and align the face\n",
    "    style_aligned_path = os.path.join('style_images_aligned', f'{name}.png')\n",
    "    if not os.path.exists(style_aligned_path):\n",
    "        style_aligned = align_face(style_path)\n",
    "        style_aligned.save(style_aligned_path)\n",
    "    else:\n",
    "        style_aligned = Image.open(style_aligned_path).convert('RGB')\n",
    "\n",
    "    # GAN invert\n",
    "    style_code_path = os.path.join('inversion_codes', f'{name}.pt')\n",
    "    if not os.path.exists(style_code_path):\n",
    "        latent = e4e_projection(style_aligned, style_code_path, device)\n",
    "    else:\n",
    "        latent = torch.load(style_code_path)['latent']\n",
    "\n",
    "    targets.append(transform(style_aligned).to(device))\n",
    "    latents.append(latent.to(device))\n",
    "\n",
    "targets = torch.stack(targets, 0)\n",
    "latents = torch.stack(latents, 0)\n",
    "\n",
    "target_im = utils.make_grid(targets, normalize=True)\n",
    "display_image(target_im, title='Style References')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "colab": {
     "base_uri": "https://localhost:8080/"
    },
    "id": "_qNPut_ch3gr",
    "outputId": "194357b6-9ace-4e9f-c003-1e002dfb1345"
   },
   "outputs": [],
   "source": [
    "#@title Finetune StyleGAN\n",
    "#@markdown alpha controls the strength of the style\n",
    "alpha =  1.0 #@param {type:\"slider\", min:0, max:1, step:0.1}\n",
    "alpha = 1-alpha\n",
    "\n",
    "#@markdown Tries to preserve color of original image by limiting family of allowable transformations. Set to false if you want to transfer color from reference image. This also leads to heavier stylization\n",
    "preserve_color = False #@param{type:\"boolean\"}\n",
    "#@markdown Number of finetuning steps. Different style reference may require different iterations. Try 200~500 iterations.\n",
    "num_iter = 300 #@param {type:\"number\"}\n",
    "#@markdown Log training on wandb and interval for image logging\n",
    "use_wandb = False #@param {type:\"boolean\"}\n",
    "log_interval = 50 #@param {type:\"number\"}\n",
    "\n",
    "if use_wandb:\n",
    "    wandb.init(project=\"JoJoGAN\")\n",
    "    config = wandb.config\n",
    "    config.num_iter = num_iter\n",
    "    config.preserve_color = preserve_color\n",
    "    wandb.log(\n",
    "    {\"Style reference\": [wandb.Image(transforms.ToPILImage()(target_im))]},\n",
    "    step=0)\n",
    "\n",
    "# load discriminator for perceptual loss\n",
    "discriminator = Discriminator(1024, 2).eval().to(device)\n",
    "ckpt = torch.load('models\\stylegan2-ffhq-config-f.pt', map_location=lambda storage, loc: storage)\n",
    "discriminator.load_state_dict(ckpt[\"d\"], strict=False)\n",
    "\n",
    "# reset generator\n",
    "del generator\n",
    "generator = deepcopy(original_generator)\n",
    "\n",
    "g_optim = optim.Adam(generator.parameters(), lr=2e-3, betas=(0, 0.99))\n",
    "\n",
    "# Which layers to swap for generating a family of plausible real images -> fake image\n",
    "if preserve_color:\n",
    "    id_swap = [9,11,15,16,17]\n",
    "else:\n",
    "    id_swap = list(range(7, generator.n_latent))\n",
    "\n",
    "for idx in tqdm(range(num_iter)):\n",
    "    mean_w = generator.get_latent(torch.randn([latents.size(0), latent_dim]).to(device)).unsqueeze(1).repeat(1, generator.n_latent, 1)\n",
    "    in_latent = latents.clone()\n",
    "    in_latent[:, id_swap] = alpha*latents[:, id_swap] + (1-alpha)*mean_w[:, id_swap]\n",
    "\n",
    "    img = generator(in_latent, input_is_latent=True)\n",
    "\n",
    "    with torch.no_grad():\n",
    "        real_feat = discriminator(targets)\n",
    "    fake_feat = discriminator(img)\n",
    "\n",
    "    loss = sum([F.l1_loss(a, b) for a, b in zip(fake_feat, real_feat)])\\len(fake_feat)\n",
    "\n",
    "    if use_wandb:\n",
    "        wandb.log({\"loss\": loss}, step=idx)\n",
    "        if idx % log_interval == 0:\n",
    "            generator.eval()\n",
    "            my_sample = generator(my_w, input_is_latent=True)\n",
    "            generator.train()\n",
    "            my_sample = transforms.ToPILImage()(utils.make_grid(my_sample, normalize=True, range=(-1, 1)))\n",
    "            wandb.log(\n",
    "            {\"Current stylization\": [wandb.Image(my_sample)]},\n",
    "            step=idx)\n",
    "\n",
    "    g_optim.zero_grad()\n",
    "    loss.backward()\n",
    "    g_optim.step()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "colab": {
     "base_uri": "https://localhost:8080/",
     "height": 1000
    },
    "id": "-zf7xz33ty_E",
    "outputId": "4e2a5f77-df48-45c3-c057-f9937ce589b8"
   },
   "outputs": [],
   "source": [
    "#@title Generate results\n",
    "n_sample =  5#@param {type:\"number\"}\n",
    "seed = 3000 #@param {type:\"number\"}\n",
    "\n",
    "torch.manual_seed(seed)\n",
    "with torch.no_grad():\n",
    "    generator.eval()\n",
    "    z = torch.randn(n_sample, latent_dim, device=device)\n",
    "\n",
    "    original_sample = original_generator([z], truncation=0.7, truncation_latent=mean_latent)\n",
    "    sample = generator([z], truncation=0.7, truncation_latent=mean_latent)\n",
    "\n",
    "    original_my_sample = original_generator(my_w, input_is_latent=True)\n",
    "    my_sample = generator(my_w, input_is_latent=True)\n",
    "\n",
    "# display reference images\n",
    "style_images = []\n",
    "for name in names:\n",
    "    style_path = f'style_images_aligned\\{strip_path_extension(name)}.png'\n",
    "    style_image = transform(Image.open(style_path))\n",
    "    style_images.append(style_image)\n",
    "\n",
    "face = transform(aligned_face).to(device).unsqueeze(0)\n",
    "style_images = torch.stack(style_images, 0).to(device)\n",
    "display_image(utils.make_grid(style_images, normalize=True), title='References')\n",
    "\n",
    "my_output = torch.cat([face, my_sample], 0)\n",
    "display_image(utils.make_grid(my_output, normalize=True), title='My sample')\n",
    "\n",
    "output = torch.cat([original_sample, sample], 0)\n",
    "display_image(utils.make_grid(output, normalize=True, nrow=n_sample), title='Random samples')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "colab": {
     "base_uri": "https://localhost:8080/",
     "height": 427
    },
    "id": "lAOQxXkgHKDF",
    "outputId": "ccf22541-726e-4bb5-805b-df7dde51ff11"
   },
   "outputs": [],
   "source": [
    "import cv2\n",
    "import torch\n",
    "import numpy as np\n",
    "from torchvision import transforms\n",
    "from copy import deepcopy\n",
    "\n",
    "# Load your models (assuming they are defined as in your notebook)\n",
    "device = 'cuda' if torch.cuda.is_available() else 'cpu'\n",
    "\n",
    "# Load the specific pre-trained model you want to use\n",
    "generator = torch.load('models\\arcane_jinx.pt')  # Example: change to the model you want\n",
    "generator = deepcopy(generator).to(device)\n",
    "generator.eval()\n",
    "\n",
    "def preprocess(frame):\n",
    "    transform = transforms.Compose([\n",
    "        transforms.ToPILImage(),\n",
    "        transforms.Resize((256, 256)),\n",
    "        transforms.ToTensor(),\n",
    "        transforms.Normalize((0.5, 0.5, 0.5), (0.5, 0.5, 0.5)),\n",
    "    ])\n",
    "    return transform(frame).unsqueeze(0).to(device)  # Add batch dimension and move to GPU\n",
    "\n",
    "def postprocess(tensor):\n",
    "    image = tensor.squeeze(0).cpu()  # Remove batch dimension and move to CPU\n",
    "    image = (image * 0.5) + 0.5  # Denormalize\n",
    "    image = transforms.ToPILImage()(image)\n",
    "    return np.array(image)\n",
    "\n",
    "# Initialize the camera\n",
    "cap = cv2.VideoCapture(0)\n",
    "\n",
    "while True:\n",
    "    ret, frame = cap.read()\n",
    "    if not ret:\n",
    "        break\n",
    "\n",
    "    frame = cv2.cvtColor(frame, cv2.COLOR_BGR2RGB)  # Convert to RGB\n",
    "    input_tensor = preprocess(frame)\n",
    "\n",
    "    with torch.no_grad():\n",
    "        stylized_tensor = generator(input_tensor, input_is_latent=True)\n",
    "\n",
    "    stylized_frame = postprocess(stylized_tensor)\n",
    "    stylized_frame = cv2.cvtColor(stylized_frame, cv2.COLOR_RGB2BGR)  # Convert back to BGR for OpenCV\n",
    "\n",
    "    cv2.imshow('Styled Frame', stylized_frame)\n",
    "\n",
    "    if cv2.waitKey(1) & 0xFF == ord('q'):  # Press 'q' to quit\n",
    "        break\n",
    "\n",
    "cap.release()\n",
    "cv2.destroyAllWindows()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "id": "3nNUvDfDNp5e"
   },
   "outputs": [],
   "source": [
    "import os\n",
    "import torch\n",
    "\n",
    "# Define the directory to save the models\n",
    "styles_folder = 'styles_change'\n",
    "os.makedirs(styles_folder, exist_ok=True)\n",
    "\n",
    "# Define the Google Drive IDs for the models\n",
    "drive_ids = {\n",
    "    \"stylegan2-ffhq-config-f.pt\": \"1Yr7KuD959btpmcKGAUsbAk5rPjX2MytK\",\n",
    "    \"e4e_ffhq_encode.pt\": \"1o6ijA3PkcewZvwJJ73dJ0fxhndn0nnh7\",\n",
    "    \"restyle_psp_ffhq_encode.pt\": \"1nbxCIVw9H3YnQsoIPykNEFwWJnHVHlVd\",\n",
    "    \"arcane_caitlyn.pt\": \"1gOsDTiTPcENiFOrhmkkxJcTURykW1dRc\",\n",
    "    \"arcane_caitlyn_preserve_color.pt\": \"1cUTyjU-q98P75a8THCaO545RTwpVV-aH\",\n",
    "    \"arcane_jinx_preserve_color.pt\": \"1jElwHxaYPod5Itdy18izJk49K1nl4ney\",\n",
    "    \"arcane_jinx.pt\": \"1quQ8vPjYpUiXM4k1_KIwP4EccOefPpG_\",\n",
    "    \"arcane_multi_preserve_color.pt\": \"1enJgrC08NpWpx2XGBmLt1laimjpGCyfl\",\n",
    "    \"arcane_multi.pt\": \"15V9s09sgaw-zhKp116VHigf5FowAy43f\",\n",
    "    \"sketch_multi.pt\": \"1GdaeHGBGjBAFsWipTL0y-ssUiAqk8AxD\",\n",
    "    \"disney.pt\": \"1zbE2upakFUAx8ximYnLofFwfT8MilqJA\",\n",
    "    \"disney_preserve_color.pt\": \"1Bnh02DjfvN_Wm8c4JdOiNV4q9J7Z_tsi\",\n",
    "    \"jojo.pt\": \"13cR2xjIBj8Ga5jMO7gtxzIJj2PDsBYK4\",\n",
    "    \"jojo_preserve_color.pt\": \"1ZRwYLRytCEKi__eT2Zxv1IlV6BGVQ_K2\",\n",
    "    \"jojo_yasuho.pt\": \"1grZT3Gz1DLzFoJchAmoj3LoM9ew9ROX_\",\n",
    "    \"jojo_yasuho_preserve_color.pt\": \"1SKBu1h0iRNyeKBnya_3BBmLr4pkPeg_L\",\n",
    "    \"art.pt\": \"1a0QDEHwXQ6hE_FcYEyNMuv5r5UnRQLKT\",\n",
    "}\n",
    "\n",
    "# Define the Downloader class\n",
    "class Downloader(object):\n",
    "    def __init__(self, use_pydrive):\n",
    "        self.use_pydrive = use_pydrive\n",
    "\n",
    "        if self.use_pydrive:\n",
    "            self.authenticate()\n",
    "\n",
    "    def authenticate(self):\n",
    "        from pydrive.auth import GoogleAuth\n",
    "        from pydrive.drive import GoogleDrive\n",
    "        from oauth2client.client import GoogleCredentials\n",
    "        auth.authenticate_user()\n",
    "        gauth = GoogleAuth()\n",
    "        gauth.credentials = GoogleCredentials.get_application_default()\n",
    "        self.drive = GoogleDrive(gauth)\n",
    "\n",
    "    def download_file(self, file_name):\n",
    "        file_dst = os.path.join(styles_folder, file_name)\n",
    "        file_id = drive_ids[file_name]\n",
    "        if not os.path.exists(file_dst):\n",
    "            print(f'Downloading {file_name}')\n",
    "            if self.use_pydrive:\n",
    "                downloaded = self.drive.CreateFile({'id':file_id})\n",
    "                downloaded.FetchMetadata(fetch_all=True)\n",
    "                downloaded.GetContentFile(file_dst)\n",
    "            else:\n",
    "                !gdown --id $file_id -O $file_dst\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "colab": {
     "base_uri": "https://localhost:8080/"
    },
    "id": "om_gGQYHOgZf",
    "outputId": "594590d3-f343-427a-9b46-3a8074d3d0d3"
   },
   "outputs": [],
   "source": [
    "# Instantiate the Downloader class and download all models\n",
    "download_with_pydrive = False  # Change to True if using pydrive\n",
    "downloader = Downloader(download_with_pydrive)\n",
    "\n",
    "# Download each model into the styles_change folder\n",
    "for model_name in drive_ids.keys():\n",
    "    downloader.download_file(model_name)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "id": "_5VqYd2gO1Ov"
   },
   "outputs": [],
   "source": [
    "import cv2\n",
    "import numpy as np\n",
    "from torchvision import transforms\n",
    "from copy import deepcopy\n",
    "\n",
    "# Function to preprocess the frame\n",
    "def preprocess(frame):\n",
    "    transform = transforms.Compose([\n",
    "        transforms.ToPILImage(),\n",
    "        transforms.Resize((256, 256)),\n",
    "        transforms.ToTensor(),\n",
    "        transforms.Normalize((0.5, 0.5, 0.5), (0.5, 0.5, 0.5)),\n",
    "    ])\n",
    "    return transform(frame).unsqueeze(0).to(device)  # Add batch dimension and move to GPU\n",
    "\n",
    "# Function to postprocess the model output\n",
    "def postprocess(tensor):\n",
    "    image = tensor.squeeze(0).cpu()  # Remove batch dimension and move to CPU\n",
    "    image = (image * 0.5) + 0.5  # Denormalize\n",
    "    image = transforms.ToPILImage()(image)\n",
    "    return np.array(image)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "colab": {
     "base_uri": "https://localhost:8080/"
    },
    "id": "XNSgVYiEPX7p",
    "outputId": "70122d45-6229-4177-9e07-061b1c8fef6a"
   },
   "outputs": [],
   "source": [
    "def apply_style_in_real_time(model_name):\n",
    "    device = 'cuda' if torch.cuda.is_available() else 'cpu'\n",
    "    model_path = os.path.join(styles_folder, model_name)\n",
    "\n",
    "    # Load the checkpoint\n",
    "    checkpoint = torch.load(model_path, map_location=device)\n",
    "\n",
    "    # Initialize the generator model\n",
    "    generator = Generator(1024, 512, 8, 2).to(device)\n",
    "\n",
    "    # Load the model weights from the correct key in the checkpoint\n",
    "    generator.load_state_dict(checkpoint[\"g\"], strict=False)  # Use the key 'g' to load weights\n",
    "    generator.eval()\n",
    "\n",
    "    # Initialize the camera\n",
    "    cap = cv2.VideoCapture(0)\n",
    "\n",
    "    while True:\n",
    "        ret, frame = cap.read()\n",
    "        if not ret:\n",
    "            break\n",
    "\n",
    "        frame = cv2.cvtColor(frame, cv2.COLOR_BGR2RGB)  # Convert to RGB\n",
    "        input_tensor = preprocess(frame)\n",
    "\n",
    "        with torch.no_grad():\n",
    "            stylized_tensor = generator(input_tensor, input_is_latent=True)\n",
    "\n",
    "        stylized_frame = postprocess(stylized_tensor)\n",
    "        stylized_frame = cv2.cvtColor(stylized_frame, cv2.COLOR_RGB2BGR)  # Convert back to BGR for OpenCV\n",
    "\n",
    "        cv2.imshow('Styled Frame', stylized_frame)\n",
    "\n",
    "        if cv2.waitKey(1) & 0xFF == ord('q'):  # Press 'q' to quit\n",
    "            break\n",
    "\n",
    "    cap.release()\n",
    "    cv2.destroyAllWindows()\n",
    "\n",
    "# Example usage: Apply a specific style model\n",
    "apply_style_in_real_time('arcane_jinx.pt')  # Replace with any model name you downloaded\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "colab": {
     "base_uri": "https://localhost:8080/"
    },
    "id": "J6lTQ6A_PbSv",
    "outputId": "8717f644-67f1-47cd-b586-fc7775df118b"
   },
   "outputs": [],
   "source": [
    "apply_style_in_real_time('arcane_jinx.pt')\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "id": "UzMsA_TcRunE"
   },
   "outputs": [],
   "source": [
    "def apply_style_in_real_time(model_name):\n",
    "    device = 'cuda' if torch.cuda.is_available() else 'cpu'\n",
    "    model_path = os.path.join(styles_folder, model_name)\n",
    "\n",
    "    # Load the checkpoint\n",
    "    checkpoint = torch.load(model_path, map_location=device)\n",
    "    print(\"Checkpoint loaded successfully.\")\n",
    "\n",
    "    # Initialize the generator model\n",
    "    generator = Generator(1024, 512, 8, 2).to(device)\n",
    "\n",
    "    # Load the model weights from the correct key in the checkpoint\n",
    "    generator.load_state_dict(checkpoint[\"g\"], strict=False)  # Use the key 'g' to load weights\n",
    "    generator.eval()\n",
    "    print(\"Model loaded and ready.\")\n",
    "\n",
    "    # Initialize the camera\n",
    "    cap = cv2.VideoCapture(0)\n",
    "    print(\"Camera initialized.\")\n",
    "\n",
    "    while True:\n",
    "        ret, frame = cap.read()\n",
    "        if not ret:\n",
    "            print(\"Failed to capture frame.\")\n",
    "            break\n",
    "\n",
    "        frame = cv2.cvtColor(frame, cv2.COLOR_BGR2RGB)  # Convert to RGB\n",
    "        input_tensor = preprocess(frame)\n",
    "\n",
    "        with torch.no_grad():\n",
    "            stylized_tensor = generator(input_tensor, input_is_latent=True)\n",
    "        print(\"Frame processed.\")\n",
    "\n",
    "        stylized_frame = postprocess(stylized_tensor)\n",
    "        stylized_frame = cv2.cvtColor(stylized_frame, cv2.COLOR_RGB2BGR)  # Convert back to BGR for OpenCV\n",
    "\n",
    "        cv2.imshow('Styled Frame', stylized_frame)\n",
    "\n",
    "        if cv2.waitKey(1) & 0xFF == ord('q'):  # Press 'q' to quit\n",
    "            print(\"Exit requested.\")\n",
    "            break\n",
    "\n",
    "    cap.release()\n",
    "    cv2.destroyAllWindows()\n",
    "    print(\"Camera released and window closed.\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "accelerator": "GPU",
  "colab": {
   "provenance": []
  },
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.12.4"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
